// @gao/view/dist/helpers.js
/**
 * @gao/view — View Helpers
 *
 * Standard helpers available in all templates.
 */
import { AssetPipeline } from './assets.js';
export class ViewHelpers {
    options;
    constructor(options = {}) {
        this.options = options;
    }
    /**
     * Generate a URL for the given path.
     */
    url(path) {
        const base = this.options.baseUrl || '';
        return `${base.replace(/\/$/, '')}/${path.replace(/^\//, '')}`;
    }
    /**
     * Get the versioned asset path.
     */
    asset(path) {
        if (this.options.assetPipeline) {
            return this.options.assetPipeline.resolve(path);
        }
        return `/${path.replace(/^\//, '')}`;
    }
    /**
     * Generate a CSRF hidden input field.
     */
    csrf() {
        const token = this.options.csrfToken || '';
        return `<input type="hidden" name="_csrf" value="${token}">`;
    }
    /**
     * Check if user has specific permission.
     */
    can(permission) {
        if (!this.options.userPermissions)
            return false;
        return this.options.userPermissions.includes(permission);
    }
    /**
     * Helper for form re-population (old value).
     */
    old(key, data, defaultValue = '') {
        if (data && data[key] !== undefined) {
            return data[key];
        }
        return defaultValue;
    }
    /**
     * Generate pagination HTML (simplified).
     */
    paginate(meta) {
        if (!meta || !meta.totalPages || meta.totalPages <= 1)
            return '';
        let html = '<nav class="pagination">';
        if (meta.hasPrev) {
            html += `<a href="?page=${meta.page - 1}">Previous</a>`;
        }
        html += `<span>Page ${meta.page} of ${meta.totalPages}</span>`;
        if (meta.hasNext) {
            html += `<a href="?page=${meta.page + 1}">Next</a>`;
        }
        html += '</nav>';
        return html;
    }
}
