// @gao/orm/dist/active-record.js
/**
 * @gao/orm — Active Record
 *
 * Provides static finders, instance persistence, and eager loading.
 * Extends BaseModel to create a full Active Record pattern.
 */
import { getMetadata } from './decorators.js';
import { executeHooks } from './hooks.js';
import { QueryBuilder } from './query-builder.js';
/** Global driver reference shared by all models */
let globalDriver = null;
let globalDialect = 'postgres';
/**
 * Set the global database driver for all Active Record models.
 * Should be called once during app bootstrap.
 */
export function setModelDriver(driver, dialect = 'postgres') {
    globalDriver = driver;
    globalDialect = dialect;
}
/**
 * Get the global driver. Throws if not initialized.
 */
export function getModelDriver() {
    if (!globalDriver) {
        throw new Error('Model driver not initialized. Call setModelDriver(driver) during app bootstrap.');
    }
    return globalDriver;
}
export function getModelDialect() {
    return globalDialect;
}
/**
 * Model — Active Record base class.
 *
 * Usage:
 * ```ts
 * @Table('contacts')
 * class Contact extends Model {
 *   @Column() name!: string;
 *   @Column() email!: string;
 * }
 *
 * const contact = await Contact.find('uuid-here');
 * contact.name = 'Updated';
 * await contact.save();
 * ```
 */
export class Model {
    /** Primary Key (UUID v4 by default) */
    id;
    created_at;
    updated_at;
    deleted_at;
    /** Tracks whether this is a new (unsaved) record */
    _isNew = true;
    /** Store original values for dirty tracking */
    _originalAttributes = {};
    /** Eagerly loaded relations */
    _relations = {};
    // ─── Static: Connection ──────────────────────────────────────
    static get driver() {
        return getModelDriver();
    }
    static get dialect() {
        return getModelDialect();
    }
    static get tableName() {
        return getMetadata(this).tableName;
    }
    // ─── Static: Query Builder ───────────────────────────────────
    /**
     * Returns a typed QueryBuilder for this model's table.
     */
    static query() {
        const meta = getMetadata(this);
        return new QueryBuilder(this.driver, this.dialect, meta.tableName);
    }
    // ─── Static: Finders ─────────────────────────────────────────
    /**
     * Find a model by its primary key.
     */
    static async find(id) {
        const meta = getMetadata(this);
        const driver = getModelDriver();
        const qb = new QueryBuilder(driver, getModelDialect(), meta.tableName);
        const rows = await qb.where('id', id).limit(1).get();
        if (rows.length === 0)
            return null;
        return this.hydrate(rows[0]);
    }
    /**
     * Find a model or throw NotFoundError.
     */
    static async findOrFail(id) {
        const result = await this.find(id);
        if (!result) {
            throw new Error(`${this.name} with id '${id}' not found`);
        }
        return result;
    }
    /**
     * Get all records (respects soft delete — excludes deleted by default).
     */
    static async all() {
        const meta = getMetadata(this);
        const driver = getModelDriver();
        const qb = new QueryBuilder(driver, getModelDialect(), meta.tableName);
        qb.whereNull('deleted_at');
        const rows = await qb.get();
        return rows.map((row) => this.hydrate(row));
    }
    /**
     * Start a WHERE chain.
     */
    static where(column, operatorOrValue, value) {
        const meta = getMetadata(this);
        const driver = getModelDriver();
        const qb = new QueryBuilder(driver, getModelDialect(), meta.tableName);
        if (value === undefined && operatorOrValue !== undefined) {
            qb.where(column, operatorOrValue);
        }
        else if (operatorOrValue !== undefined) {
            qb.where(column, operatorOrValue, value);
        }
        return new ModelQueryBuilder(qb, this);
    }
    /**
     * Create a new model, save it to the database, and return it.
     */
    static async create(data) {
        const instance = new this();
        instance.fill(data);
        await instance.save();
        return instance;
    }
    /**
     * Find by match criteria or create a new record.
     */
    static async updateOrCreate(match, data) {
        const meta = getMetadata(this);
        const driver = getModelDriver();
        const qb = new QueryBuilder(driver, getModelDialect(), meta.tableName);
        for (const [key, val] of Object.entries(match)) {
            qb.where(key, val);
        }
        const rows = await qb.limit(1).get();
        if (rows.length > 0) {
            const instance = this.hydrate(rows[0]);
            instance.fill(data);
            await instance.save();
            return instance;
        }
        return this.create({ ...match, ...data });
    }
    // ─── Static: Eager Loading ───────────────────────────────────
    /**
     * Start a query with eager loading.
     */
    static with(...relations) {
        const meta = getMetadata(this);
        const driver = getModelDriver();
        const qb = new QueryBuilder(driver, getModelDialect(), meta.tableName);
        const mqb = new ModelQueryBuilder(qb, this);
        mqb.eagerLoad(...relations);
        return mqb;
    }
    // ─── Static: Hydration ───────────────────────────────────────
    /**
     * Create a model instance from a raw database row.
     */
    static hydrate(row) {
        const instance = Object.create(this.prototype);
        // Assign all row values to instance
        for (const [key, value] of Object.entries(row)) {
            instance[key] = value;
        }
        instance._isNew = false;
        instance._originalAttributes = { ...row };
        instance._relations = {};
        return instance;
    }
    // ─── Instance: Persistence ───────────────────────────────────
    /**
     * Fill model with data.
     */
    fill(data) {
        for (const [key, value] of Object.entries(data)) {
            if (key !== '_isNew' && key !== '_originalAttributes' && key !== '_relations') {
                this[key] = value;
            }
        }
        return this;
    }
    /**
     * Save the model — INSERT if new, UPDATE dirty fields if existing.
     */
    async save() {
        const meta = getMetadata(this.constructor);
        const driver = getModelDriver();
        const dialect = getModelDialect();
        const tableName = meta.tableName;
        // Run BeforeSave hooks
        await executeHooks(this, 'beforeSave');
        if (this._isNew) {
            // INSERT
            if (!this.id) {
                const { randomUUID } = await import('node:crypto');
                this.id = randomUUID();
            }
            this.created_at = this.created_at || new Date().toISOString();
            this.updated_at = new Date().toISOString();
            const attributes = this.getAttributes();
            const qb = new QueryBuilder(driver, dialect, tableName);
            await qb.insert(attributes);
            this._isNew = false;
        }
        else {
            // UPDATE — only dirty fields
            const dirty = this.getDirty();
            if (Object.keys(dirty).length === 0)
                return this;
            dirty.updated_at = new Date().toISOString();
            this.updated_at = dirty.updated_at;
            const qb = new QueryBuilder(driver, dialect, tableName);
            await qb.where('id', this.id).update(dirty);
        }
        this.syncOriginal();
        // Run AfterSave hooks
        await executeHooks(this, 'afterSave');
        return this;
    }
    /**
     * Destroy the model — soft delete by default, hard delete with force: true.
     */
    async destroy(options) {
        const meta = getMetadata(this.constructor);
        const driver = getModelDriver();
        const dialect = getModelDialect();
        const tableName = meta.tableName;
        // Run BeforeDelete hooks
        await executeHooks(this, 'beforeDelete');
        const qb = new QueryBuilder(driver, dialect, tableName);
        if (options?.force) {
            // Hard delete
            await qb.where('id', this.id).delete();
        }
        else {
            // Soft delete
            this.deleted_at = new Date().toISOString();
            await qb.where('id', this.id).update({ deleted_at: this.deleted_at });
        }
        // Run AfterDelete hooks
        await executeHooks(this, 'afterDelete');
    }
    /**
     * Refresh the model from the database.
     */
    async refresh() {
        const meta = getMetadata(this.constructor);
        const driver = getModelDriver();
        const dialect = getModelDialect();
        const tableName = meta.tableName;
        const qb = new QueryBuilder(driver, dialect, tableName);
        const rows = await qb.where('id', this.id).limit(1).get();
        if (rows.length === 0) {
            throw new Error(`Model ${this.constructor.name} with id '${this.id}' no longer exists`);
        }
        const row = rows[0];
        for (const [key, value] of Object.entries(row)) {
            this[key] = value;
        }
        this.syncOriginal();
        return this;
    }
    // ─── Instance: Attribute Tracking ────────────────────────────
    getAttributes() {
        const attributes = {};
        for (const key of Object.keys(this)) {
            if (key.startsWith('_') || typeof this[key] === 'function')
                continue;
            attributes[key] = this[key];
        }
        return attributes;
    }
    syncOriginal() {
        this._originalAttributes = { ...this.getAttributes() };
        return this;
    }
    isDirty(attribute) {
        const dirty = this.getDirty();
        if (attribute)
            return Object.prototype.hasOwnProperty.call(dirty, attribute);
        return Object.keys(dirty).length > 0;
    }
    getDirty() {
        const current = this.getAttributes();
        const dirty = {};
        for (const [key, value] of Object.entries(current)) {
            if (this._originalAttributes[key] !== value) {
                dirty[key] = value;
            }
        }
        return dirty;
    }
    // ─── Instance: Serialization ─────────────────────────────────
    toJSON() {
        const attrs = this.getAttributes();
        // Include eagerly loaded relations
        for (const [key, value] of Object.entries(this._relations)) {
            attrs[key] = value;
        }
        return attrs;
    }
    // ─── Instance: Relation Access ───────────────────────────────
    /**
     * Get eagerly loaded relation data.
     */
    getRelation(name) {
        return this._relations[name];
    }
    /**
     * Set relation data (used by eager loader).
     */
    setRelation(name, value) {
        this._relations[name] = value;
    }
}
// ─── ModelQueryBuilder ─────────────────────────────────────────
/**
 * A typed wrapper around QueryBuilder that returns hydrated Model instances.
 */
export class ModelQueryBuilder {
    qb;
    ModelClass;
    _eagerLoads = [];
    constructor(qb, ModelClass) {
        this.qb = qb;
        this.ModelClass = ModelClass;
    }
    // ─── Chainable WHERE methods ─────────────────────────────────
    where(column, operatorOrValue, value) {
        if (value === undefined) {
            this.qb.where(column, operatorOrValue);
        }
        else {
            this.qb.where(column, operatorOrValue, value);
        }
        return this;
    }
    orWhere(column, operatorOrValue, value) {
        if (value === undefined) {
            this.qb.orWhere(column, operatorOrValue);
        }
        else {
            this.qb.orWhere(column, operatorOrValue, value);
        }
        return this;
    }
    whereIn(column, values) {
        this.qb.whereIn(column, values);
        return this;
    }
    whereNotIn(column, values) {
        this.qb.whereNotIn(column, values);
        return this;
    }
    whereNull(column) {
        this.qb.whereNull(column);
        return this;
    }
    whereNotNull(column) {
        this.qb.whereNotNull(column);
        return this;
    }
    whereBetween(column, range) {
        this.qb.whereBetween(column, range);
        return this;
    }
    // ─── Chainable other methods ─────────────────────────────────
    orderBy(column, direction = 'ASC') {
        this.qb.orderBy(column, direction);
        return this;
    }
    limit(limit) {
        this.qb.limit(limit);
        return this;
    }
    offset(offset) {
        this.qb.offset(offset);
        return this;
    }
    select(...columns) {
        this.qb.select(...columns);
        return this;
    }
    groupBy(...columns) {
        this.qb.groupBy(...columns);
        return this;
    }
    // ─── Eager Loading ───────────────────────────────────────────
    eagerLoad(...relations) {
        this._eagerLoads.push(...relations);
        return this;
    }
    with(...relations) {
        return this.eagerLoad(...relations);
    }
    // ─── Soft Delete Scopes ──────────────────────────────────────
    withDeleted() {
        // Don't add the whereNull('deleted_at') filter
        return this;
    }
    onlyDeleted() {
        this.qb.whereNotNull('deleted_at');
        return this;
    }
    // ─── Execution ───────────────────────────────────────────────
    /**
     * Get all matching records as hydrated model instances.
     */
    async get() {
        const rows = await this.qb.get();
        const instances = rows.map((row) => this.ModelClass.hydrate(row));
        // Process eager loads
        if (this._eagerLoads.length > 0 && instances.length > 0) {
            await this.processEagerLoads(instances);
        }
        return instances;
    }
    /**
     * Get the first matching record.
     */
    async first() {
        this.qb.limit(1);
        const instances = await this.get();
        return instances[0] ?? null;
    }
    /**
     * Paginate results.
     */
    async paginate(page = 1, perPage = 15) {
        const result = await this.qb.paginate(page, perPage);
        const instances = result.data.map((row) => this.ModelClass.hydrate(row));
        if (this._eagerLoads.length > 0 && instances.length > 0) {
            await this.processEagerLoads(instances);
        }
        return {
            data: instances,
            meta: result.meta,
        };
    }
    /**
     * Count matching records.
     */
    async count(column = '*') {
        return this.qb.count(column);
    }
    /**
     * Check if matching records exist.
     */
    async exists() {
        return this.qb.exists();
    }
    /**
     * Get single column values.
     */
    async pluck(column) {
        return this.qb.pluck(column);
    }
    // ─── Private: Eager Loading ──────────────────────────────────
    async processEagerLoads(instances) {
        for (const relationName of this._eagerLoads) {
            const instance = instances[0];
            const proto = Object.getPrototypeOf(instance);
            if (typeof proto[relationName] !== 'function') {
                continue; // Skip if relation method doesn't exist
            }
            // Get the relation definition
            const relation = proto[relationName].call(instance);
            // Collect all local key values for batch loading
            const localValues = instances.map((inst) => inst[relation.localKey]);
            const uniqueValues = [...new Set(localValues)];
            if (uniqueValues.length === 0)
                continue;
            // Batch query: fetch all related records in one query
            const meta = getMetadata(relation['foreignModel']);
            const driver = getModelDriver();
            const relQb = new QueryBuilder(driver, getModelDialect(), meta.tableName);
            relQb.whereIn(relation.foreignKey, uniqueValues);
            const relatedRows = await relQb.get();
            // Group related records by foreign key
            const grouped = new Map();
            for (const row of relatedRows) {
                const key = row[relation.foreignKey];
                if (!grouped.has(key))
                    grouped.set(key, []);
                grouped.get(key).push(row);
            }
            // Assign to each instance
            for (const inst of instances) {
                const localValue = inst[relation.localKey];
                const related = grouped.get(localValue) || [];
                // Determine if HasOne/BelongsTo (single) or HasMany/ManyToMany (array)
                const relationType = relation.constructor.name;
                if (relationType === 'HasOne' || relationType === 'BelongsTo') {
                    inst.setRelation(relationName, related[0] || null);
                }
                else {
                    inst.setRelation(relationName, related);
                }
            }
        }
    }
}
