// @gao/orm/dist/connection.js
/**
 * @gao/orm — Connection Manager
 *
 * Manages database driver selection, initialization, health checks,
 * and graceful shutdown operations.
 */
export class ConnectionManager {
    config;
    driver = null;
    isConnected = false;
    constructor(config) {
        this.config = config;
    }
    /**
     * Set the compiled driver manually. Helpful for dependency injection.
     */
    setDriver(driver) {
        this.driver = driver;
    }
    /**
     * Connects to the database using the configured driver.
     * Implements a retry-loop for health checking the connection.
     */
    async connect() {
        if (!this.driver) {
            throw new Error(`Database driver not defined. Configured for type: ${this.config.driver}`);
        }
        if (this.isConnected)
            return;
        const retries = this.config.healthRetryCount ?? 3;
        const delay = this.config.healthRetryDelayMs ?? 1000;
        let attempt = 1;
        while (attempt <= retries) {
            try {
                await this.driver.connect();
                // Perform a simple health check query (valid for both PG and SQLite)
                await this.driver.query('SELECT 1 as healthy');
                this.isConnected = true;
                return;
            }
            catch (error) {
                if (attempt === retries) {
                    throw new Error(`Database connection failed after ${retries} attempts: ${error.message}`);
                }
                await new Promise((res) => setTimeout(res, delay));
                attempt++;
            }
        }
    }
    /**
     * Returns the active database driver instance.
     */
    getDriver() {
        if (!this.driver || !this.isConnected) {
            throw new Error('Database is not connected. Call connect() first.');
        }
        return this.driver;
    }
    /**
     * Gracefully disconnects all connections.
     */
    async disconnect() {
        if (!this.driver || !this.isConnected)
            return;
        await this.driver.disconnect();
        this.isConnected = false;
    }
}
