// @gao/orm/dist/decorators.js
/**
 * @gao/orm — Model Decorators
 *
 * Provides metadata routing for models using TypeScript decorators.
 */
export const ORM_METADATA_KEY = Symbol('gao:orm:metadata');
/**
 * Retrieves or initializes the metadata for a model class.
 */
export function getMetadata(target) {
    if (!target[ORM_METADATA_KEY]) {
        target[ORM_METADATA_KEY] = {
            tableName: target.name.toLowerCase() + 's',
            columns: {},
            indexes: {},
            uniques: [],
            foreignKeys: {},
            encryptedFields: [],
            hooks: {
                beforeSave: [],
                afterSave: [],
                beforeDelete: [],
                afterDelete: [],
            },
            scopes: {},
        };
    }
    return target[ORM_METADATA_KEY];
}
/**
 * Class decorator to define the table name.
 */
export function Table(name) {
    return ((target) => {
        const meta = getMetadata(target);
        meta.tableName = name;
    });
}
/**
 * Property decorator to define a column.
 */
export function Column(options = {}) {
    return (target, propertyKey) => {
        const meta = getMetadata(target.constructor);
        meta.columns[propertyKey] = options;
    };
}
/**
 * Property decorator to define an index.
 */
export function Index(options = {}) {
    return (target, propertyKey) => {
        const meta = getMetadata(target.constructor);
        meta.indexes[propertyKey] = options;
    };
}
/**
 * Property decorator to mark a column as unique.
 */
export function Unique() {
    return (target, propertyKey) => {
        const meta = getMetadata(target.constructor);
        meta.uniques.push(propertyKey);
    };
}
/**
 * Property decorator to define a foreign key relationship.
 */
export function ForeignKey(options) {
    return (target, propertyKey) => {
        const meta = getMetadata(target.constructor);
        meta.foreignKeys[propertyKey] = options;
    };
}
/**
 * Property decorator to mark a field for auto-encryption.
 */
export function Encrypted() {
    return (target, propertyKey) => {
        const meta = getMetadata(target.constructor);
        meta.encryptedFields.push(propertyKey);
    };
}
