// @gao/orm/dist/model.js
/**
 * @gao/orm — BaseModel
 *
 * Abstract base class for all entity models.
 * Provides UUID primary key, timestamp handling, serialization, and dirty attribute tracking.
 */
import { randomUUID } from 'node:crypto';
export class BaseModel {
    /** Primary Key (UUID v4) */
    id;
    /** Created timestamp, auto-managed */
    created_at;
    /** Updated timestamp, auto-managed */
    updated_at;
    /** Deleted timestamp for soft-delete functionality */
    deleted_at;
    /** Store original values to track dirty fields */
    _originalAttributes = {};
    constructor(initialData = {}) {
        this.id = randomUUID();
        this.fill(initialData);
        this.syncOriginal();
    }
    /**
     * Populate model attributes securely.
     */
    fill(data) {
        for (const [key, value] of Object.entries(data)) {
            // In a real active record, we'd check fillable/guarded properties.
            // For BaseModel, we just assign.
            // @ts-ignore - Dynamic assignment
            this[key] = value;
        }
        return this;
    }
    /**
     * Synchronize the current state as the original, unmutated state.
     * Should be called after loading from the database or after a successful save.
     */
    syncOriginal() {
        this._originalAttributes = JSON.parse(JSON.stringify(this.getAttributes()));
        return this;
    }
    /**
     * Retrieve a plain object of the model's attributes.
     */
    getAttributes() {
        const attributes = {};
        for (const key of Object.keys(this)) {
            if (key !== '_originalAttributes' &&
                typeof this[key] !== 'function') {
                attributes[key] = this[key];
            }
        }
        return attributes;
    }
    /**
     * Check if the model has been modified since it was last synced.
     * @param attribute Optional specific attribute to check
     */
    isDirty(attribute) {
        const dirty = this.getDirty();
        if (attribute) {
            return Object.prototype.hasOwnProperty.call(dirty, attribute);
        }
        return Object.keys(dirty).length > 0;
    }
    /**
     * Get an object of all attributes that have changed.
     */
    getDirty() {
        const current = this.getAttributes();
        const dirty = {};
        for (const [key, value] of Object.entries(current)) {
            if (this._originalAttributes[key] !== value) {
                // Simplistic equality check. In production, deep equality may be needed.
                dirty[key] = value;
            }
        }
        return dirty;
    }
    /**
     * Soft delete the model (sets deleted_at timestamp).
     */
    delete() {
        this.deleted_at = new Date().toISOString();
    }
    /**
     * Restore a soft-deleted model.
     */
    restore() {
        this.deleted_at = undefined;
    }
    /**
     * Serialize model for HTTP response. Masks hidden fields if implemented.
     */
    toJSON() {
        return this.getAttributes();
    }
}
