// @gao/orm/dist/query-builder.js
/**
 * @gao/orm — Query Builder
 *
 * Fluent API for constructing type-safe parameterized SQL queries.
 * Prevents SQL injection by strictly using parameterized queries.
 */
export class QueryBuilder {
    driver;
    dialect;
    _selects = ['*'];
    _rawSelects = [];
    _from = '';
    _wheres = [];
    _joins = [];
    _orderBys = [];
    _groupBys = [];
    _havings = [];
    _limit;
    _offset;
    constructor(driver, dialect = 'postgres', tableName) {
        this.driver = driver;
        this.dialect = dialect;
        if (tableName) {
            this._from = tableName;
        }
    }
    from(table) {
        this._from = table;
        return this;
    }
    select(...columns) {
        if (columns.length > 0) {
            this._selects = columns;
        }
        return this;
    }
    where(column, operatorOrValue, value) {
        if (value === undefined) {
            this._wheres.push({ column, operator: '=', value: operatorOrValue, boolean: 'AND' });
        }
        else {
            this._wheres.push({ column, operator: operatorOrValue, value, boolean: 'AND' });
        }
        return this;
    }
    orWhere(column, operatorOrValue, value) {
        if (value === undefined) {
            this._wheres.push({ column, operator: '=', value: operatorOrValue, boolean: 'OR' });
        }
        else {
            this._wheres.push({ column, operator: operatorOrValue, value, boolean: 'OR' });
        }
        return this;
    }
    join(table, first, operator, second, type = 'INNER') {
        this._joins.push({ table, first, operator, second, type });
        return this;
    }
    leftJoin(table, first, operator, second) {
        return this.join(table, first, operator, second, 'LEFT');
    }
    rightJoin(table, first, operator, second) {
        return this.join(table, first, operator, second, 'RIGHT');
    }
    // ─── Extended WHERE Methods ────────────────────────────────────
    whereIn(column, values) {
        if (values.length === 0) {
            // Empty array — always false
            this._wheres.push({ column: '1', operator: '=', value: 0, boolean: 'AND', raw: true });
            return this;
        }
        this._wheres.push({ column, operator: 'IN', value: values, boolean: 'AND' });
        return this;
    }
    whereNotIn(column, values) {
        if (values.length === 0)
            return this; // NOT IN empty = no restriction
        this._wheres.push({ column, operator: 'NOT IN', value: values, boolean: 'AND' });
        return this;
    }
    whereNull(column) {
        this._wheres.push({ column, operator: 'IS', value: null, boolean: 'AND' });
        return this;
    }
    whereNotNull(column) {
        this._wheres.push({ column, operator: 'IS NOT', value: null, boolean: 'AND' });
        return this;
    }
    whereBetween(column, range) {
        this._wheres.push({ column, operator: 'BETWEEN', value: range, boolean: 'AND' });
        return this;
    }
    whereRaw(sql, params = []) {
        this._wheres.push({ column: sql, operator: '=', value: params, boolean: 'AND', raw: true });
        return this;
    }
    orWhereIn(column, values) {
        if (values.length === 0)
            return this;
        this._wheres.push({ column, operator: 'IN', value: values, boolean: 'OR' });
        return this;
    }
    // ─── Extended SELECT Methods ──────────────────────────────────
    selectRaw(...expressions) {
        this._rawSelects.push(...expressions);
        return this;
    }
    // ─── GROUP BY / HAVING ────────────────────────────────────────
    groupBy(...columns) {
        this._groupBys.push(...columns);
        return this;
    }
    having(column, operatorOrValue, value) {
        if (value === undefined) {
            this._havings.push({ column, operator: '=', value: operatorOrValue });
        }
        else {
            this._havings.push({ column, operator: operatorOrValue, value });
        }
        return this;
    }
    orderBy(column, direction = 'ASC') {
        this._orderBys.push({ column, direction });
        return this;
    }
    limit(limit) {
        this._limit = limit;
        return this;
    }
    offset(offset) {
        this._offset = offset;
        return this;
    }
    /**
     * Compile a SELECT query.
     */
    toSql() {
        const params = [];
        // Build SELECT clause
        const selectParts = [];
        if (this._rawSelects.length > 0) {
            selectParts.push(...this._rawSelects);
            if (this._selects.length > 0 && !(this._selects.length === 1 && this._selects[0] === '*')) {
                selectParts.push(...this._selects);
            }
        }
        else {
            selectParts.push(...this._selects);
        }
        let sql = `SELECT ${selectParts.join(', ')} FROM ${this.quoteIdentifier(this._from)}`;
        for (const join of this._joins) {
            sql += ` ${join.type} JOIN ${this.quoteIdentifier(join.table)} ON ${this.quoteIdentifier(join.first)} ${join.operator} ${this.quoteIdentifier(join.second)}`;
        }
        sql += this.compileWheres(params);
        if (this._groupBys.length > 0) {
            const groupSql = this._groupBys.map((col) => this.quoteIdentifier(col)).join(', ');
            sql += ` GROUP BY ${groupSql}`;
        }
        if (this._havings.length > 0) {
            sql += ' HAVING ';
            this._havings.forEach((having, index) => {
                if (index > 0)
                    sql += ' AND ';
                sql += `${this.quoteIdentifier(having.column)} ${having.operator} ${this.formatParam(params.length)}`;
                params.push(having.value);
            });
        }
        if (this._orderBys.length > 0) {
            const orderSql = this._orderBys
                .map((ob) => `${this.quoteIdentifier(ob.column)} ${ob.direction}`)
                .join(', ');
            sql += ` ORDER BY ${orderSql}`;
        }
        if (this._limit !== undefined) {
            sql += ` LIMIT ${this._limit}`;
        }
        if (this._offset !== undefined) {
            sql += ` OFFSET ${this._offset}`;
        }
        return { sql, params };
    }
    /**
     * Compile WHERE clauses into SQL string fragment and push params.
     */
    compileWheres(params) {
        if (this._wheres.length === 0)
            return '';
        let sql = ' WHERE ';
        this._wheres.forEach((where, index) => {
            if (index > 0)
                sql += ` ${where.boolean} `;
            if (where.raw) {
                // whereRaw — inject SQL directly, append params
                sql += where.column;
                if (Array.isArray(where.value) && where.value.length > 0) {
                    params.push(...where.value);
                }
                return;
            }
            if (where.operator === 'IN' || where.operator === 'NOT IN') {
                const values = where.value;
                const placeholders = values.map((_, i) => this.formatParam(params.length + i)).join(', ');
                sql += `${this.quoteIdentifier(where.column)} ${where.operator} (${placeholders})`;
                params.push(...values);
                return;
            }
            if (where.operator === 'IS' || where.operator === 'IS NOT') {
                sql += `${this.quoteIdentifier(where.column)} ${where.operator} NULL`;
                return;
            }
            if (where.operator === 'BETWEEN') {
                const range = where.value;
                sql += `${this.quoteIdentifier(where.column)} BETWEEN ${this.formatParam(params.length)} AND ${this.formatParam(params.length + 1)}`;
                params.push(range[0], range[1]);
                return;
            }
            sql += `${this.quoteIdentifier(where.column)} ${where.operator} ${this.formatParam(params.length)}`;
            params.push(where.value);
        });
        return sql;
    }
    /**
     * Setup an INSERT query using the provided data.
     */
    compileInsert(data) {
        const columns = Object.keys(data);
        const params = Object.values(data);
        const placeholders = columns.map((_, i) => this.formatParam(i)).join(', ');
        const colsSql = columns.map((c) => this.quoteIdentifier(c)).join(', ');
        const sql = `INSERT INTO ${this.quoteIdentifier(this._from)} (${colsSql}) VALUES (${placeholders})`;
        return { sql, params };
    }
    /**
     * Setup an UPDATE query.
     */
    compileUpdate(data) {
        const params = [];
        const setSql = Object.entries(data)
            .map(([col, val]) => {
            params.push(val);
            return `${this.quoteIdentifier(col)} = ${this.formatParam(params.length - 1)}`;
        })
            .join(', ');
        let sql = `UPDATE ${this.quoteIdentifier(this._from)} SET ${setSql}`;
        sql += this.compileWheres(params);
        return { sql, params };
    }
    /**
     * Setup a DELETE query.
     */
    compileDelete() {
        const params = [];
        let sql = `DELETE FROM ${this.quoteIdentifier(this._from)}`;
        sql += this.compileWheres(params);
        return { sql, params };
    }
    /**
     * Compile a batch INSERT query.
     */
    compileInsertMany(rows) {
        if (rows.length === 0)
            throw new Error('insertMany requires at least one row');
        const columns = Object.keys(rows[0]);
        const params = [];
        const valueSets = [];
        for (const row of rows) {
            const placeholders = columns.map((col) => {
                params.push(row[col]);
                return this.formatParam(params.length - 1);
            });
            valueSets.push(`(${placeholders.join(', ')})`);
        }
        const colsSql = columns.map((c) => this.quoteIdentifier(c)).join(', ');
        const sql = `INSERT INTO ${this.quoteIdentifier(this._from)} (${colsSql}) VALUES ${valueSets.join(', ')}`;
        return { sql, params };
    }
    /**
     * Compile an UPSERT query (INSERT ... ON CONFLICT).
     */
    compileUpsert(data, conflictColumns, updateColumns) {
        const { sql: insertSql, params } = this.compileInsert(data);
        const conflictCols = conflictColumns.map((c) => this.quoteIdentifier(c)).join(', ');
        const updateCols = updateColumns || Object.keys(data).filter((k) => !conflictColumns.includes(k));
        const updateSql = updateCols
            .map((col) => `${this.quoteIdentifier(col)} = EXCLUDED.${this.quoteIdentifier(col)}`)
            .join(', ');
        const sql = updateSql
            ? `${insertSql} ON CONFLICT (${conflictCols}) DO UPDATE SET ${updateSql}`
            : `${insertSql} ON CONFLICT (${conflictCols}) DO NOTHING`;
        return { sql, params };
    }
    // ─── Execute Methods ──────────────────────────────────────────
    async get() {
        const { sql, params } = this.toSql();
        return this.driver.query(sql, params);
    }
    async first() {
        this.limit(1);
        const results = await this.get();
        return results[0] ?? null;
    }
    async paginate(page = 1, perPage = 15) {
        const _originalSelects = this._selects;
        const _originalLimit = this._limit;
        const _originalOffset = this._offset;
        const _originalOrderBys = this._orderBys;
        // Count query
        this._selects = ['COUNT(*) as total'];
        this._limit = undefined;
        this._offset = undefined;
        this._orderBys = []; // Order by doesn't affect count
        const { sql: countSql, params: countParams } = this.toSql();
        const countResult = await this.driver.query(countSql, countParams);
        const totalRows = countResult[0]?.total ? Number(countResult[0]?.total) : 0;
        // Restore builder state
        this._selects = _originalSelects;
        this._limit = _originalLimit;
        this._offset = _originalOffset;
        this._orderBys = _originalOrderBys;
        // Actual query
        const currentPage = Math.max(1, page);
        const limit = Math.max(1, perPage);
        const offset = (currentPage - 1) * limit;
        this.limit(limit);
        this.offset(offset);
        const data = await this.get();
        const totalPages = Math.ceil(totalRows / limit);
        return {
            data,
            meta: {
                page: currentPage,
                perPage: limit,
                total: totalRows,
                totalPages,
                hasPrev: currentPage > 1,
                hasNext: currentPage < totalPages,
            },
        };
    }
    async insert(data) {
        const { sql, params } = this.compileInsert(data);
        const finalSql = this.dialect === 'postgres' ? `${sql} RETURNING id` : sql;
        return this.driver.execute(finalSql, params);
    }
    async insertMany(rows) {
        const { sql, params } = this.compileInsertMany(rows);
        const finalSql = this.dialect === 'postgres' ? `${sql} RETURNING id` : sql;
        return this.driver.execute(finalSql, params);
    }
    async upsert(data, conflictColumns, updateColumns) {
        const { sql, params } = this.compileUpsert(data, conflictColumns, updateColumns);
        const finalSql = this.dialect === 'postgres' ? `${sql} RETURNING id` : sql;
        return this.driver.execute(finalSql, params);
    }
    async update(data) {
        const { sql, params } = this.compileUpdate(data);
        return this.driver.execute(sql, params);
    }
    async increment(column, amount = 1) {
        const params = [amount];
        let sql = `UPDATE ${this.quoteIdentifier(this._from)} SET ${this.quoteIdentifier(column)} = ${this.quoteIdentifier(column)} + ${this.formatParam(0)}`;
        sql += this.compileWheres(params);
        return this.driver.execute(sql, params);
    }
    async decrement(column, amount = 1) {
        const params = [amount];
        let sql = `UPDATE ${this.quoteIdentifier(this._from)} SET ${this.quoteIdentifier(column)} = ${this.quoteIdentifier(column)} - ${this.formatParam(0)}`;
        sql += this.compileWheres(params);
        return this.driver.execute(sql, params);
    }
    async delete() {
        const { sql, params } = this.compileDelete();
        return this.driver.execute(sql, params);
    }
    // ─── Aggregate Methods ────────────────────────────────────────
    async count(column = '*') {
        const originalSelects = this._selects;
        const originalRawSelects = this._rawSelects;
        this._selects = [];
        this._rawSelects = [`COUNT(${column === '*' ? '*' : this.quoteIdentifier(column)}) as "aggregate"`];
        const { sql, params } = this.toSql();
        this._selects = originalSelects;
        this._rawSelects = originalRawSelects;
        const rows = await this.driver.query(sql, params);
        return Number(rows[0]?.aggregate ?? 0);
    }
    async sum(column) {
        const originalSelects = this._selects;
        const originalRawSelects = this._rawSelects;
        this._selects = [];
        this._rawSelects = [`COALESCE(SUM(${this.quoteIdentifier(column)}), 0) as "aggregate"`];
        const { sql, params } = this.toSql();
        this._selects = originalSelects;
        this._rawSelects = originalRawSelects;
        const rows = await this.driver.query(sql, params);
        return Number(rows[0]?.aggregate ?? 0);
    }
    async avg(column) {
        const originalSelects = this._selects;
        const originalRawSelects = this._rawSelects;
        this._selects = [];
        this._rawSelects = [`COALESCE(AVG(${this.quoteIdentifier(column)}), 0) as "aggregate"`];
        const { sql, params } = this.toSql();
        this._selects = originalSelects;
        this._rawSelects = originalRawSelects;
        const rows = await this.driver.query(sql, params);
        return Number(rows[0]?.aggregate ?? 0);
    }
    async min(column) {
        const originalSelects = this._selects;
        const originalRawSelects = this._rawSelects;
        this._selects = [];
        this._rawSelects = [`MIN(${this.quoteIdentifier(column)}) as "aggregate"`];
        const { sql, params } = this.toSql();
        this._selects = originalSelects;
        this._rawSelects = originalRawSelects;
        const rows = await this.driver.query(sql, params);
        return Number(rows[0]?.aggregate ?? 0);
    }
    async max(column) {
        const originalSelects = this._selects;
        const originalRawSelects = this._rawSelects;
        this._selects = [];
        this._rawSelects = [`MAX(${this.quoteIdentifier(column)}) as "aggregate"`];
        const { sql, params } = this.toSql();
        this._selects = originalSelects;
        this._rawSelects = originalRawSelects;
        const rows = await this.driver.query(sql, params);
        return Number(rows[0]?.aggregate ?? 0);
    }
    async pluck(column) {
        const originalSelects = this._selects;
        this._selects = [this.quoteIdentifier(column)];
        const { sql, params } = this.toSql();
        this._selects = originalSelects;
        const rows = await this.driver.query(sql, params);
        return rows.map((row) => row[column]);
    }
    async exists() {
        const originalSelects = this._selects;
        const originalRawSelects = this._rawSelects;
        const originalLimit = this._limit;
        this._selects = ['1'];
        this._rawSelects = [];
        this._limit = 1;
        const { sql, params } = this.toSql();
        this._selects = originalSelects;
        this._rawSelects = originalRawSelects;
        this._limit = originalLimit;
        const rows = await this.driver.query(sql, params);
        return rows.length > 0;
    }
    // ─── Helpers ──────────────────────────────────────────────────
    quoteIdentifier(identifier) {
        // Protect from naive SQL injection in identifiers.
        // Identifiers shouldn't contain spaces or quotes unless properly escaped.
        // For simplicity, we wrap in double quotes.
        if (identifier.includes('.')) {
            return identifier
                .split('.')
                .map((part) => `"${part}"`)
                .join('.');
        }
        return `"${identifier}"`;
    }
    formatParam(index) {
        if (this.dialect === 'postgres') {
            return `$${index + 1}`;
        }
        return '?';
    }
}
