// @gao/orm/dist/relations.js
/**
 * @gao/orm — Model Relations
 *
 * Provides relation types: HasOne, HasMany, BelongsTo, ManyToMany.
 */
import { getMetadata } from './decorators.js';
import { QueryBuilder } from './query-builder.js';
export class Relation {
    foreignModel;
    localKey;
    foreignKey;
    constructor(foreignModel, localKey, foreignKey) {
        this.foreignModel = foreignModel;
        this.localKey = localKey;
        this.foreignKey = foreignKey;
    }
    getTableName(model) {
        return getMetadata(model).tableName || model.name.toLowerCase() + 's';
    }
}
export class HasOne extends Relation {
    query(driver, localValue) {
        const tableName = this.getTableName(this.foreignModel);
        return new QueryBuilder(driver, 'postgres', tableName)
            .where(this.foreignKey, '=', localValue)
            .limit(1);
    }
}
export class HasMany extends Relation {
    query(driver, localValue) {
        const tableName = this.getTableName(this.foreignModel);
        return new QueryBuilder(driver, 'postgres', tableName).where(this.foreignKey, '=', localValue);
    }
}
export class BelongsTo extends Relation {
    query(driver, localValue) {
        const tableName = this.getTableName(this.foreignModel);
        return new QueryBuilder(driver, 'postgres', tableName)
            .where(this.localKey, '=', localValue) // Here localKey points to the PK of foreign model
            .limit(1);
    }
}
export class ManyToMany extends Relation {
    pivotTable;
    pivotLocalKey;
    pivotForeignKey;
    constructor(foreignModel, localKey, foreignKey, pivotTable, pivotLocalKey, pivotForeignKey) {
        super(foreignModel, localKey, foreignKey);
        this.pivotTable = pivotTable;
        this.pivotLocalKey = pivotLocalKey;
        this.pivotForeignKey = pivotForeignKey;
    }
    query(driver, localValue) {
        const foreignTable = this.getTableName(this.foreignModel);
        // SELECT * FROM foreignTable INNER JOIN pivotTable ON foreignTable.foreignKey = pivotTable.pivotForeignKey WHERE pivotTable.pivotLocalKey = ?
        return new QueryBuilder(driver, 'postgres', foreignTable)
            .join(this.pivotTable, `${foreignTable}.${this.foreignKey}`, '=', `${this.pivotTable}.${this.pivotForeignKey}`)
            .where(`${this.pivotTable}.${this.pivotLocalKey}`, '=', localValue);
    }
}
