// @gao/orm/dist/schema.js
/**
 * @gao/orm — Schema Builder
 *
 * Fluent API for generating SQL DDL statements.
 */
export class ColumnBuilder {
    name;
    type;
    length;
    precision;
    scale;
    isPrimary = false;
    isNullable = false;
    isUnique = false;
    defaultValue;
    constructor(name, type, length, precision, scale) {
        this.name = name;
        this.type = type;
        this.length = length;
        this.precision = precision;
        this.scale = scale;
    }
    primary() {
        this.isPrimary = true;
        return this;
    }
    nullable() {
        this.isNullable = true;
        return this;
    }
    unique() {
        this.isUnique = true;
        return this;
    }
    default(val) {
        if (typeof val === 'string' &&
            val !== 'CURRENT_TIMESTAMP' &&
            val !== 'NOW()' &&
            val !== 'CURRENT_DATE' &&
            val !== 'CURRENT_TIME') {
            this.defaultValue = `'${val}'`;
        }
        else {
            this.defaultValue = String(val);
        }
        return this;
    }
    compile() {
        let sql = `"${this.name}" `;
        switch (this.type) {
            case 'varchar':
                sql += `VARCHAR(${this.length ?? 255})`;
                break;
            case 'uuid':
                sql += 'UUID';
                break;
            case 'text':
                sql += 'TEXT';
                break;
            case 'integer':
                sql += 'INTEGER';
                break;
            case 'bigint':
                sql += 'BIGINT';
                break;
            case 'boolean':
                sql += 'BOOLEAN';
                break;
            case 'timestamp':
                sql += 'TIMESTAMP';
                break;
            case 'decimal':
            case 'numeric': {
                const p = this.precision ?? 10;
                const s = this.scale ?? 2;
                sql += `${this.type.toUpperCase()}(${p}, ${s})`;
                break;
            }
            case 'json':
                sql += 'JSON';
                break;
            case 'jsonb':
                sql += 'JSONB';
                break;
            case 'date':
                sql += 'DATE';
                break;
            case 'time':
                sql += 'TIME';
                break;
            case 'serial':
                sql += 'SERIAL';
                break;
            case 'bigserial':
                sql += 'BIGSERIAL';
                break;
        }
        if (this.isPrimary)
            sql += ' PRIMARY KEY';
        if (!this.isNullable && !this.isPrimary && this.type !== 'serial' && this.type !== 'bigserial')
            sql += ' NOT NULL';
        if (this.isUnique)
            sql += ' UNIQUE';
        if (this.defaultValue !== undefined)
            sql += ` DEFAULT ${this.defaultValue}`;
        return sql;
    }
}
export class TableBuilder {
    columns = [];
    foreignKeys = [];
    indexes = [];
    checks = [];
    uuid(name) {
        const col = new ColumnBuilder(name, 'uuid');
        this.columns.push(col);
        return col;
    }
    string(name, length = 255) {
        const col = new ColumnBuilder(name, 'varchar', length);
        this.columns.push(col);
        return col;
    }
    text(name) {
        const col = new ColumnBuilder(name, 'text');
        this.columns.push(col);
        return col;
    }
    integer(name) {
        const col = new ColumnBuilder(name, 'integer');
        this.columns.push(col);
        return col;
    }
    bigint(name) {
        const col = new ColumnBuilder(name, 'bigint');
        this.columns.push(col);
        return col;
    }
    boolean(name) {
        const col = new ColumnBuilder(name, 'boolean');
        this.columns.push(col);
        return col;
    }
    timestamp(name) {
        const col = new ColumnBuilder(name, 'timestamp');
        this.columns.push(col);
        return col;
    }
    decimal(name, precision = 10, scale = 2) {
        const col = new ColumnBuilder(name, 'decimal', undefined, precision, scale);
        this.columns.push(col);
        return col;
    }
    numeric(name, precision = 10, scale = 2) {
        const col = new ColumnBuilder(name, 'numeric', undefined, precision, scale);
        this.columns.push(col);
        return col;
    }
    json(name) {
        const col = new ColumnBuilder(name, 'json');
        this.columns.push(col);
        return col;
    }
    jsonb(name) {
        const col = new ColumnBuilder(name, 'jsonb');
        this.columns.push(col);
        return col;
    }
    date(name) {
        const col = new ColumnBuilder(name, 'date');
        this.columns.push(col);
        return col;
    }
    time(name) {
        const col = new ColumnBuilder(name, 'time');
        this.columns.push(col);
        return col;
    }
    serial(name) {
        const col = new ColumnBuilder(name, 'serial');
        this.columns.push(col);
        return col;
    }
    bigserial(name) {
        const col = new ColumnBuilder(name, 'bigserial');
        this.columns.push(col);
        return col;
    }
    enum(name, values) {
        this.checks.push({
            expression: `"${name}" IN (${values.map((v) => `'${v}'`).join(', ')})`,
            name: `chk_${name}_enum`,
        });
        const col = new ColumnBuilder(name, 'varchar', 255);
        this.columns.push(col);
        return col;
    }
    timestamps() {
        this.timestamp('created_at').default('CURRENT_TIMESTAMP');
        this.timestamp('updated_at').nullable();
        this.timestamp('deleted_at').nullable();
    }
    foreignKey(column, referenceTable, referenceColumn = 'id', onDelete = 'CASCADE', onUpdate = 'NO ACTION') {
        this.foreignKeys.push({ column, referenceTable, referenceColumn, onDelete, onUpdate });
    }
    index(columns, name, unique = false) {
        const cols = Array.isArray(columns) ? columns : [columns];
        const indexName = name || `idx_${cols.join('_')}`;
        this.indexes.push({ name: indexName, columns: cols, unique });
    }
    uniqueIndex(columns, name) {
        this.index(columns, name, true);
    }
    check(expression, name) {
        this.checks.push({ expression, name });
    }
}
export class AlterTableBuilder {
    tableName;
    operations = [];
    constructor(tableName) {
        this.tableName = tableName;
    }
    addColumn(builder) {
        this.operations.push(`ADD COLUMN ${builder.compile()}`);
        return this;
    }
    dropColumn(name) {
        this.operations.push(`DROP COLUMN "${name}"`);
        return this;
    }
    renameColumn(oldName, newName) {
        this.operations.push(`RENAME COLUMN "${oldName}" TO "${newName}"`);
        return this;
    }
    addForeignKey(column, referenceTable, referenceColumn = 'id', onDelete = 'CASCADE') {
        this.operations.push(`ADD CONSTRAINT "fk_${this.tableName}_${column}" FOREIGN KEY ("${column}") REFERENCES "${referenceTable}"("${referenceColumn}") ON DELETE ${onDelete}`);
        return this;
    }
    dropConstraint(name) {
        this.operations.push(`DROP CONSTRAINT "${name}"`);
        return this;
    }
    compile() {
        return this.operations.map((op) => `ALTER TABLE "${this.tableName}" ${op};`);
    }
}
// biome-ignore lint/complexity/noStaticOnlyClass: Schema acts as a namespace for builder
export class Schema {
    static create(tableName, callback) {
        const builder = new TableBuilder();
        callback(builder);
        const parts = [];
        // Column definitions
        parts.push(...builder.columns.map((c) => c.compile()));
        // Foreign keys
        for (const fk of builder.foreignKeys) {
            parts.push(`CONSTRAINT "fk_${tableName}_${fk.column}" FOREIGN KEY ("${fk.column}") REFERENCES "${fk.referenceTable}"("${fk.referenceColumn}") ON DELETE ${fk.onDelete} ON UPDATE ${fk.onUpdate}`);
        }
        // Check constraints
        for (const chk of builder.checks) {
            const name = chk.name ? `CONSTRAINT "${chk.name}" ` : '';
            parts.push(`${name}CHECK (${chk.expression})`);
        }
        let sql = `CREATE TABLE "${tableName}" (\n  ${parts.join(',\n  ')}\n);`;
        // Index statements (separate from CREATE TABLE)
        for (const idx of builder.indexes) {
            const uniqueStr = idx.unique ? 'UNIQUE ' : '';
            const colsSql = idx.columns.map((c) => `"${c}"`).join(', ');
            sql += `\nCREATE ${uniqueStr}INDEX "${idx.name}" ON "${tableName}" (${colsSql});`;
        }
        return sql;
    }
    static drop(tableName) {
        return `DROP TABLE IF EXISTS "${tableName}";`;
    }
    static alter(tableName, callback) {
        const builder = new AlterTableBuilder(tableName);
        callback(builder);
        return builder.compile();
    }
    static dropIfExists(tableName) {
        return `DROP TABLE IF EXISTS "${tableName}" CASCADE;`;
    }
    static rename(oldName, newName) {
        return `ALTER TABLE "${oldName}" RENAME TO "${newName}";`;
    }
}
