// @gao/core/dist/scoped-container.js
/**
 * @gao/core — Scoped Container
 *
 * A restricted view of the DI Container that enforces permission checks.
 * Plugins receive a ScopedContainer instead of the full Container,
 * ensuring they can only access services they've declared in their permissions.
 */
import { ForbiddenError } from './errors.js';
import { PermissionSet, SERVICE_PERMISSIONS } from './permission.js';
export class ScopedContainer {
    parent;
    permissions;
    pluginName;
    constructor(parent, permissions, pluginName) {
        this.parent = parent;
        this.permissions = permissions;
        this.pluginName = pluginName;
    }
    /**
     * Resolve a service by its token.
     * Checks if the plugin has permission to access the service.
     * If the service has no registered permission requirement, access is granted.
     *
     * @throws ForbiddenError if the plugin lacks the required permission.
     */
    resolve(token) {
        const requiredPermission = SERVICE_PERMISSIONS.get(token);
        if (requiredPermission && !this.permissions.has(requiredPermission)) {
            throw new ForbiddenError(`Plugin "${this.pluginName}" denied access to "${String(token)}". ` +
                `Required permission: "${requiredPermission}". ` +
                `Declared permissions: ${this.permissions.toString()}`);
        }
        return this.parent.resolve(token);
    }
    /**
     * Check if a service is registered AND the plugin has permission to access it.
     */
    has(token) {
        const requiredPermission = SERVICE_PERMISSIONS.get(token);
        if (requiredPermission && !this.permissions.has(requiredPermission)) {
            return false;
        }
        return this.parent.has(token);
    }
    /** Get the plugin name this scope belongs to. */
    get owner() {
        return this.pluginName;
    }
    /** Get the permission set. */
    get grantedPermissions() {
        return this.permissions;
    }
}
/**
 * Create a scoped container for a plugin with specific permissions.
 */
export function createScopedContainer(parent, pluginName, permissions) {
    return new ScopedContainer(parent, new PermissionSet(permissions), pluginName);
}
