// @gao/security/dist/middleware/validator.js
/**
 * @gao/security — Zod Validator Middleware
 *
 * Auto-validates request body, params, and query against Zod schemas.
 * Returns structured validation errors.
 */
import { ValidationError } from '@gao/core';
import { z } from 'zod';
/**
 * Zod validation middleware wrapper.
 */
export function validate(options) {
    const { body, query, params, stripUnknown = true } = options;
    return {
        name: 'security:validator',
        // biome-ignore lint/complexity/noExcessiveCognitiveComplexity: Middleware handle functions are inherently procedural
        handle: async (ctx, next) => {
            // In a real adapter, request object contains these
            const request = ctx.metadata.request;
            if (!request)
                return next();
            const errors = [];
            // Validate body
            if (body) {
                const result = await body.safeParseAsync(request.body);
                if (!result.success) {
                    errors.push(...result.error.issues.map((i) => ({
                        field: `body.${i.path.join('.')}`,
                        message: i.message,
                        code: i.code,
                    })));
                }
                else {
                    // Replace with stripped/transformed data
                    request.body =
                        stripUnknown && body instanceof z.ZodObject
                            ? body.strip().parse(request.body)
                            : result.data;
                }
            }
            // Validate query
            if (query) {
                const result = await query.safeParseAsync(request.query);
                if (!result.success) {
                    errors.push(...result.error.issues.map((i) => ({
                        field: `query.${i.path.join('.')}`,
                        message: i.message,
                        code: i.code,
                    })));
                }
                else {
                    request.query =
                        stripUnknown && query instanceof z.ZodObject
                            ? query.strip().parse(request.query)
                            : result.data;
                }
            }
            // Validate params
            if (params) {
                const result = await params.safeParseAsync(request.params);
                if (!result.success) {
                    errors.push(...result.error.issues.map((i) => ({
                        field: `params.${i.path.join('.')}`,
                        message: i.message,
                        code: i.code,
                    })));
                }
                else {
                    request.params =
                        stripUnknown && params instanceof z.ZodObject
                            ? params.strip().parse(request.params)
                            : result.data;
                }
            }
            if (errors.length > 0) {
                throw new ValidationError('Request validation failed', errors);
            }
            await next();
        },
    };
}
