// @gao/security/dist/middleware/rate-limiter.js
/**
 * @gao/security — Rate Limiter Middleware
 *
 * Configurable rate limiting with support for:
 * - Specific points (requests) per duration
 * - Block duration when limit exceeded
 * - Store adapters (Memory default)
 * - Custom key generation (IP, user ID, route)
 */
import { RateLimitError } from '@gao/core';
/**
 * Default in-memory store for rate limiting.
 * Cleans up expired entries periodically to prevent memory leaks.
 */
export class MemoryRateLimiterStore {
    counts = new Map();
    blocks = new Map(); // key -> expiresAt
    cleanupInterval;
    constructor() {
        this.cleanupInterval = setInterval(() => this.cleanup(), 60_000);
        if (this.cleanupInterval.unref)
            this.cleanupInterval.unref();
    }
    async increment(key, durationMs) {
        const now = Date.now();
        let entry = this.counts.get(key);
        if (!entry || entry.expiresAt <= now) {
            entry = { count: 0, expiresAt: now + durationMs };
            this.counts.set(key, entry);
        }
        entry.count += 1;
        return { count: entry.count, resetMs: Math.max(0, entry.expiresAt - now) };
    }
    async block(key, durationMs) {
        this.blocks.set(key, Date.now() + durationMs);
    }
    async isBlocked(key) {
        const blockExpiresAt = this.blocks.get(key);
        if (!blockExpiresAt)
            return 0;
        const now = Date.now();
        if (blockExpiresAt <= now) {
            this.blocks.delete(key);
            return 0;
        }
        return blockExpiresAt - now;
    }
    cleanup() {
        const now = Date.now();
        for (const [key, entry] of this.counts.entries()) {
            if (entry.expiresAt <= now)
                this.counts.delete(key);
        }
        for (const [key, expiresAt] of this.blocks.entries()) {
            if (expiresAt <= now)
                this.blocks.delete(key);
        }
    }
    destroy() {
        clearInterval(this.cleanupInterval);
        this.counts.clear();
        this.blocks.clear();
    }
}
/**
 * Default key generator creating a key based on IP.
 */
function defaultKeyGenerator(ctx) {
    const request = ctx.metadata.request;
    const ip = request?.ip ?? 'unknown-ip';
    const path = request?.path ?? 'unknown-path';
    return `${ip}:${path}`;
}
/**
 * Rate Limiter Middleware.
 */
export function rateLimiter(options = {}) {
    const points = options.points ?? 60;
    const durationMs = (options.duration ?? 60) * 1000;
    const blockDurationMs = (options.blockDuration ?? 0) * 1000;
    const store = options.store ?? new MemoryRateLimiterStore();
    const keyGenerator = options.keyGenerator ?? defaultKeyGenerator;
    return {
        name: 'security:rate-limiter',
        handle: async (ctx, next) => {
            if (options.skip && (await options.skip(ctx))) {
                return next();
            }
            const key = await keyGenerator(ctx);
            // 1. Check if blocked
            const blockTimeRemaining = await store.isBlocked(key);
            if (blockTimeRemaining > 0) {
                throw new RateLimitError('Too many requests, blocked.', Math.ceil(blockTimeRemaining / 1000));
            }
            // 2. Increment
            const { count, resetMs } = await store.increment(key, durationMs);
            // Add rate limit headers
            let headers = ctx.metadata.responseHeaders;
            if (!headers) {
                headers = {};
                ctx.metadata.responseHeaders = headers;
            }
            headers['X-RateLimit-Limit'] = points.toString();
            headers['X-RateLimit-Remaining'] = Math.max(0, points - count).toString();
            headers['X-RateLimit-Reset'] = Math.ceil(Date.now() / 1000 + resetMs / 1000).toString();
            // 3. Check limit
            if (count > points) {
                if (blockDurationMs > 0) {
                    await store.block(key, blockDurationMs);
                    throw new RateLimitError('Too many requests, applying penalty block.', Math.ceil(blockDurationMs / 1000));
                }
                throw new RateLimitError('Too many requests.', Math.ceil(resetMs / 1000));
            }
            await next();
        },
    };
}
