// @gao/security/dist/auth/jwt.js
/**
 * @gao/security — JWT Authentication Service
 *
 * Provides access token and refresh token generation, verification, and revocation.
 * Uses `jose` library for robust cryptographic signing (RS256, HS256).
 */
import { SignJWT, jwtVerify } from 'jose';
export class MemoryRevocationAdapter {
    revokelist = new Map();
    async revoke(jti, expiresAtMs) {
        this.revokelist.set(jti, expiresAtMs);
        // Cleanup expired entries
        const now = Date.now();
        for (const [id, exp] of this.revokelist) {
            if (exp <= now)
                this.revokelist.delete(id);
        }
    }
    async isRevoked(jti) {
        const expiresAt = this.revokelist.get(jti);
        if (!expiresAt)
            return false;
        if (expiresAt <= Date.now()) {
            this.revokelist.delete(jti);
            return false;
        }
        return true;
    }
}
export class JwtService {
    config;
    revocation;
    constructor(config, revocation = new MemoryRevocationAdapter()) {
        this.config = {
            issuer: 'gao-auth',
            audience: 'gao-app',
            accessTokenTtl: '15m',
            refreshTokenTtl: '7d',
            algorithm: 'HS256',
            ...config,
        };
        this.revocation = revocation;
    }
    /**
     * Generates a pair of access and refresh tokens.
     */
    async generateTokens(subjectId, payload = {}) {
        const crypto = await import('node:crypto');
        // Create random JTI (JWT ID) for revocation support
        const accessTokenJti = crypto.randomUUID();
        const refreshTokenJti = crypto.randomUUID();
        const accessToken = await new SignJWT({ ...payload, type: 'access' })
            .setProtectedHeader({ alg: this.config.algorithm })
            .setIssuedAt()
            .setIssuer(this.config.issuer)
            .setAudience(this.config.audience)
            .setSubject(subjectId)
            .setJti(accessTokenJti)
            .setExpirationTime(this.config.accessTokenTtl)
            .sign(this.config.secretKey);
        const refreshToken = await new SignJWT({ type: 'refresh' })
            .setProtectedHeader({ alg: this.config.algorithm })
            .setIssuedAt()
            .setIssuer(this.config.issuer)
            .setAudience(this.config.audience)
            .setSubject(subjectId)
            .setJti(refreshTokenJti)
            .setExpirationTime(this.config.refreshTokenTtl)
            .sign(this.config.secretKey);
        return { accessToken, refreshToken };
    }
    /**
     * Verifies a token and extracts its payload. Throws on invalid/expired/revoked tokens.
     */
    async verify(token, expectedType = 'access') {
        const { payload } = await jwtVerify(token, this.config.secretKey, {
            issuer: this.config.issuer,
            audience: this.config.audience,
        });
        if (payload.type !== expectedType) {
            throw new Error(`Invalid token type. Expected ${expectedType}, got ${payload.type}`);
        }
        if (payload.jti && (await this.revocation.isRevoked(payload.jti))) {
            throw new Error('Token has been revoked');
        }
        return payload;
    }
    /**
     * Revoke a specific token immediately using its JTI.
     */
    async revokeToken(token) {
        try {
            // Decode without verification just to extract JTI and EXP.
            // If we used verify(), we wouldn't be able to revoke an expired token (which is fine anyway).
            const { payload } = await jwtVerify(token, this.config.secretKey, {
                issuer: this.config.issuer,
                audience: this.config.audience,
            });
            if (payload.jti && payload.exp) {
                await this.revocation.revoke(payload.jti, payload.exp * 1000);
            }
        }
        catch {
            // Ignore invalid tokens during revocation
        }
    }
    /**
     * Check if a token is valid (doesn't throw).
     */
    async isValid(token, expectedType = 'access') {
        try {
            await this.verify(token, expectedType);
            return true;
        }
        catch {
            return false;
        }
    }
}
