// @gao/orm/dist/drivers/mariadb.js
/**
 * @gao/orm — MariaDB Driver
 *
 * MariaDB driver implementation using `mariadb`.
 */
import * as mariadb from 'mariadb';
export class MariaDBDriver {
    config;
    pool = null;
    connection = null;
    isTransaction = false;
    savepointDepth = 0;
    constructor(config, connection, savepointDepth = 0) {
        this.config = config;
        if (connection) {
            this.connection = connection;
            this.isTransaction = true;
            this.savepointDepth = savepointDepth;
        }
    }
    async connect() {
        if (this.pool || this.connection)
            return;
        this.pool = mariadb.createPool(this.config);
        // Test connection
        const conn = await this.pool.getConnection();
        await conn.release();
    }
    async disconnect() {
        if (this.isTransaction)
            return;
        if (this.pool) {
            await this.pool.end();
            this.pool = null;
        }
    }
    async query(sql, params = []) {
        this.ensureConnected();
        let rows;
        if (this.isTransaction && this.connection) {
            rows = await this.connection.query(sql, params);
        }
        else {
            rows = await this.pool.query(sql, params);
        }
        // mariadb returns the array directly, unlike mysql2 which returns [rows, fields]
        return rows;
    }
    async execute(sql, params = []) {
        this.ensureConnected();
        let result;
        if (this.isTransaction && this.connection) {
            result = await this.connection.query(sql, params);
        }
        else {
            result = await this.pool.query(sql, params);
        }
        return {
            insertId: result.insertId !== undefined ? Number(result.insertId) : undefined,
            rowsAffected: result.affectedRows ?? 0,
        };
    }
    async transaction(callback) {
        this.ensureConnected();
        if (this.isTransaction && this.connection) {
            const savepointName = `sp_${this.savepointDepth + 1}`;
            await this.connection.query(`SAVEPOINT ${savepointName}`);
            try {
                const trxDriver = new MariaDBDriver(this.config, this.connection, this.savepointDepth + 1);
                const result = await callback(trxDriver);
                await this.connection.query(`RELEASE SAVEPOINT ${savepointName}`);
                return result;
            }
            catch (error) {
                await this.connection.query(`ROLLBACK TO SAVEPOINT ${savepointName}`);
                throw error;
            }
        }
        const connection = await this.pool.getConnection();
        try {
            await connection.beginTransaction();
            const trxDriver = new MariaDBDriver(this.config, connection);
            const result = await callback(trxDriver);
            await connection.commit();
            return result;
        }
        catch (error) {
            await connection.rollback();
            throw error;
        }
        finally {
            await connection.release();
        }
    }
    getPool() {
        return this.pool;
    }
    ensureConnected() {
        if (!this.pool && !this.connection) {
            throw new Error('Database not connected. Call connect() first.');
        }
    }
}
