// @gao/orm/dist/drivers/mysql.js
/**
 * @gao/orm — MySQL Driver
 *
 * MySQL driver implementation using `mysql2`. Supports prepared statements,
 * connection pooling, and nested transactions via savepoints.
 */
import mysql from 'mysql2/promise';
export class MySQLDriver {
    config;
    pool = null;
    connection = null;
    isTransaction = false;
    savepointDepth = 0;
    constructor(config, connection, savepointDepth = 0) {
        this.config = config;
        if (connection) {
            this.connection = connection;
            this.isTransaction = true;
            this.savepointDepth = savepointDepth;
        }
    }
    async connect() {
        if (this.pool || this.connection)
            return;
        if (this.config.uri) {
            this.pool = mysql.createPool(this.config.uri);
        }
        else {
            this.pool = mysql.createPool(this.config);
        }
        // Test connection
        const conn = await this.pool.getConnection();
        conn.release();
    }
    async disconnect() {
        if (this.isTransaction)
            return;
        if (this.pool) {
            await this.pool.end();
            this.pool = null;
        }
    }
    async query(sql, params = []) {
        this.ensureConnected();
        let rows;
        if (this.isTransaction && this.connection) {
            [rows] = await this.connection.execute(sql, params);
        }
        else {
            [rows] = await this.pool.execute(sql, params);
        }
        return rows;
    }
    async execute(sql, params = []) {
        this.ensureConnected();
        let result;
        if (this.isTransaction && this.connection) {
            [result] = await this.connection.execute(sql, params);
        }
        else {
            [result] = await this.pool.execute(sql, params);
        }
        const header = result;
        return {
            insertId: header.insertId,
            rowsAffected: header.affectedRows,
        };
    }
    async transaction(callback) {
        this.ensureConnected();
        if (this.isTransaction && this.connection) {
            const savepointName = `sp_${this.savepointDepth + 1}`;
            await this.connection.query(`SAVEPOINT ${savepointName}`);
            try {
                const trxDriver = new MySQLDriver(this.config, this.connection, this.savepointDepth + 1);
                const result = await callback(trxDriver);
                await this.connection.query(`RELEASE SAVEPOINT ${savepointName}`);
                return result;
            }
            catch (error) {
                await this.connection.query(`ROLLBACK TO SAVEPOINT ${savepointName}`);
                throw error;
            }
        }
        const connection = await this.pool.getConnection();
        try {
            await connection.beginTransaction();
            const trxDriver = new MySQLDriver(this.config, connection);
            const result = await callback(trxDriver);
            await connection.commit();
            return result;
        }
        catch (error) {
            await connection.rollback();
            throw error;
        }
        finally {
            connection.release();
        }
    }
    getPool() {
        return this.pool;
    }
    ensureConnected() {
        if (!this.pool && !this.connection) {
            throw new Error('Database not connected. Call connect() first.');
        }
    }
}
