// @gao/orm/dist/drivers/postgres.js
/**
 * @gao/orm — PostgreSQL Driver
 *
 * PostgreSQL driver implementation. Uses `pg` under the hood
 * to provide a high-performance asynchronous interface connected via a connection pool.
 */
import pkg from 'pg';
const { Pool } = pkg;
export class PostgresDriver {
    config;
    pool = null;
    client = null;
    isTransaction = false;
    savepointDepth = 0;
    constructor(config, client, savepointDepth = 0) {
        this.config = config;
        if (client) {
            this.client = client;
            this.isTransaction = true;
            this.savepointDepth = savepointDepth;
        }
    }
    async connect() {
        if (this.pool || this.client)
            return; // Already connected or we are a transaction driver
        this.pool = new Pool(this.config);
        // Ensure connection works
        const client = await this.pool.connect();
        client.release();
    }
    async disconnect() {
        if (this.isTransaction)
            return; // Managed by the transaction wrapper
        if (this.pool) {
            await this.pool.end();
            this.pool = null;
        }
    }
    async query(sql, params = []) {
        this.ensureConnected();
        // Parameterizing for Postgres requires $1, $2, etc instead of ?.
        // This translation is assumed to be handled by the query builder before calling driver,
        // or the user should use $1 for native manual queries. We will just pass the SQL as is.
        // It's the ORM QueryBuilder’s responsibility to output the correct dialect placeholders.
        if (this.isTransaction && this.client) {
            const { rows } = await this.client.query(sql, params);
            return rows;
        }
        const { rows } = await this.pool.query(sql, params);
        return rows;
    }
    async execute(sql, params = []) {
        this.ensureConnected();
        let result;
        if (this.isTransaction && this.client) {
            result = await this.client.query(sql, params);
        }
        else {
            result = await this.pool.query(sql, params);
        }
        // Handle insertId. In Postgres standard INSERT returning ID is the convention (`RETURNING id`).
        // If the query returns rows, we take the first assumed ID.
        let insertId = undefined;
        if (result.rows.length > 0 && result.rows[0]) {
            // Usually we'd map this explicitly if RETURNING is used. Try to infer an id if it exists.
            insertId = result.rows[0].id;
        }
        return {
            insertId,
            rowsAffected: result.rowCount ?? 0,
        };
    }
    async transaction(callback) {
        this.ensureConnected();
        if (this.isTransaction) {
            // Nested transaction -> Use Savepoints
            const savepointName = `sp_${this.savepointDepth + 1}`;
            await this.client.query(`SAVEPOINT ${savepointName}`);
            try {
                const trxDriver = new PostgresDriver(this.config, this.client, this.savepointDepth + 1);
                const result = await callback(trxDriver);
                await this.client.query(`RELEASE SAVEPOINT ${savepointName}`);
                return result;
            }
            catch (error) {
                await this.client.query(`ROLLBACK TO SAVEPOINT ${savepointName}`);
                throw error;
            }
        }
        const client = await this.pool.connect();
        try {
            await client.query('BEGIN');
            // Create a scoped driver using the specific checked out client
            const trxDriver = new PostgresDriver(this.config, client);
            const result = await callback(trxDriver);
            await client.query('COMMIT');
            return result;
        }
        catch (error) {
            await client.query('ROLLBACK');
            throw error;
        }
        finally {
            client.release();
        }
    }
    getPool() {
        return this.pool;
    }
    ensureConnected() {
        if (!this.pool && !this.client) {
            throw new Error('Database not connected. Call connect() first.');
        }
    }
}
