// @gao/security/dist/middleware/xss-guard.js
/**
 * @gao/security — XSS Guard Middleware
 *
 * Protects against Cross-Site Scripting (XSS) attacks:
 * - HTML output encoding via escape helpers
 * - HTML sanitization using DOMPurify (server-side via jsdom)
 * - CSP nonce generation for inline scripts
 * - @SanitizeInput() decorator for model fields
 */
import { randomBytes } from 'node:crypto';
let purifyInstance = null;
async function getSanitizer() {
    if (purifyInstance)
        return purifyInstance;
    const { JSDOM } = await import('jsdom');
    const dompurifyModule = await import('dompurify');
    const createDOMPurify = dompurifyModule.default;
    const window = new JSDOM('').window;
    purifyInstance = createDOMPurify(window);
    return purifyInstance;
}
/**
 * Encode special characters to HTML entities.
 * Prevents XSS in text content (not inside HTML attributes or scripts).
 */
export function htmlEncode(input) {
    return input
        .replace(/&/g, '&amp;')
        .replace(/</g, '&lt;')
        .replace(/>/g, '&gt;')
        .replace(/"/g, '&quot;')
        .replace(/'/g, '&#x27;')
        .replace(/\//g, '&#x2F;');
}
/**
 * Server-side HTML sanitization via DOMPurify.
 * Safely strips dangerous tags and attributes.
 */
export async function sanitizeHtml(dirty, allowedTags) {
    const purify = await getSanitizer();
    const config = {};
    if (allowedTags) {
        config.ALLOWED_TAGS = allowedTags;
    }
    return purify.sanitize(dirty, config);
}
/**
 * Generate a cryptographically secure nonce for CSP inline scripts.
 * Usage in HTML: <script nonce="VALUE">...</script>
 */
export function generateCspNonce() {
    return randomBytes(16).toString('base64url');
}
/**
 * XSS Guard middleware.
 * Optionally sanitizes request body and injects a CSP nonce for template use.
 */
export function xssGuard(options = {}) {
    const { sanitizeBody = false, allowedTags = [], injectCspNonce = true } = options;
    return {
        name: 'security:xss-guard',
        handle: async (ctx, next) => {
            // 1. Generate and inject CSP nonce
            if (injectCspNonce) {
                const nonce = generateCspNonce();
                ctx.metadata.cspNonce = nonce;
                // Update CSP header to include nonce
                const responseHeaders = (ctx.metadata.responseHeaders ??= {});
                const existingCsp = responseHeaders['Content-Security-Policy'];
                if (existingCsp) {
                    responseHeaders['Content-Security-Policy'] = existingCsp.replace("script-src 'self'", `script-src 'self' 'nonce-${nonce}'`);
                }
            }
            // 2. Optionally sanitize request body
            if (sanitizeBody) {
                const request = ctx.metadata.request;
                if (request?.body && typeof request.body === 'object') {
                    request.body = await sanitizeObjectStrings(request.body, allowedTags);
                }
            }
            await next();
        },
    };
}
/**
 * Recursively sanitize string values in an object.
 */
async function sanitizeObjectStrings(obj, allowedTags) {
    const result = {};
    for (const [key, value] of Object.entries(obj)) {
        if (typeof value === 'string') {
            result[key] = await sanitizeHtml(value, allowedTags);
        }
        else if (value !== null && typeof value === 'object' && !Array.isArray(value)) {
            result[key] = await sanitizeObjectStrings(value, allowedTags);
        }
        else if (Array.isArray(value)) {
            result[key] = await Promise.all(value.map((item) => typeof item === 'string'
                ? sanitizeHtml(item, allowedTags)
                : typeof item === 'object' && item !== null
                    ? sanitizeObjectStrings(item, allowedTags)
                    : Promise.resolve(item)));
        }
        else {
            result[key] = value;
        }
    }
    return result;
}
