// @gao/core/dist/lru-cache.js
/**
 * @gao/core — LRU Cache Adapter
 *
 * Memory-bounded cache with Least Recently Used eviction policy.
 * Tracks approximate byte size of stored entries, evicting the
 * least-recently-used entries when the budget is exceeded.
 *
 * Uses Map insertion-order semantics for O(1) LRU refresh.
 */
export class LRUCacheAdapter {
    store = new Map();
    currentBytes = 0;
    maxBytes;
    defaultTtl;
    cleanupTimer;
    _hitCount = 0;
    _missCount = 0;
    _evictionCount = 0;
    constructor(options = {}) {
        this.maxBytes = options.maxBytes ?? 50 * 1024 * 1024; // 50 MB
        this.defaultTtl = options.defaultTtl ?? 0;
        // Periodic expired entry cleanup
        this.cleanupTimer = setInterval(() => this.cleanupExpired(), 60_000);
        if (this.cleanupTimer.unref)
            this.cleanupTimer.unref();
    }
    async get(key) {
        const entry = this.store.get(key);
        if (!entry) {
            this._missCount++;
            return undefined;
        }
        // Check TTL
        if (entry.expiresAt !== undefined && Date.now() > entry.expiresAt) {
            this.deleteEntry(key, entry);
            this._missCount++;
            return undefined;
        }
        // LRU refresh: delete and re-insert to move to end (most recent)
        this.store.delete(key);
        this.store.set(key, entry);
        this._hitCount++;
        return entry.value;
    }
    async set(key, value, ttl) {
        const serialized = JSON.stringify(value);
        const sizeBytes = serialized.length * 2; // Rough UTF-16 estimate
        // Remove old entry if exists
        const existing = this.store.get(key);
        if (existing) {
            this.currentBytes -= existing.sizeBytes;
            this.store.delete(key);
        }
        // Evict LRU entries until under budget
        while (this.currentBytes + sizeBytes > this.maxBytes && this.store.size > 0) {
            const oldestKey = this.store.keys().next().value;
            const oldestEntry = this.store.get(oldestKey);
            this.deleteEntry(oldestKey, oldestEntry);
            this._evictionCount++;
        }
        const effectiveTtl = ttl ?? this.defaultTtl;
        const expiresAt = effectiveTtl > 0 ? Date.now() + effectiveTtl * 1000 : undefined;
        this.store.set(key, { value, sizeBytes, expiresAt });
        this.currentBytes += sizeBytes;
    }
    async delete(key) {
        const entry = this.store.get(key);
        if (!entry)
            return false;
        this.deleteEntry(key, entry);
        return true;
    }
    async clear(namespace) {
        if (!namespace) {
            this.store.clear();
            this.currentBytes = 0;
            return;
        }
        const prefix = `${namespace}:`;
        for (const [key, entry] of this.store) {
            if (key.startsWith(prefix)) {
                this.deleteEntry(key, entry);
            }
        }
    }
    async has(key) {
        const entry = this.store.get(key);
        if (!entry)
            return false;
        if (entry.expiresAt !== undefined && Date.now() > entry.expiresAt) {
            this.deleteEntry(key, entry);
            return false;
        }
        return true;
    }
    /** Number of entries in the cache. */
    get size() {
        return this.store.size;
    }
    /** Approximate bytes used by cached entries. */
    get usedBytes() {
        return this.currentBytes;
    }
    /** Maximum byte budget. */
    get budgetBytes() {
        return this.maxBytes;
    }
    /** Cache hit count. */
    get hitCount() {
        return this._hitCount;
    }
    /** Cache miss count. */
    get missCount() {
        return this._missCount;
    }
    /** Number of evictions performed. */
    get evictionCount() {
        return this._evictionCount;
    }
    /** Hit ratio: 1.0 = all hits, 0.0 = all misses. */
    get hitRatio() {
        const total = this._hitCount + this._missCount;
        if (total === 0)
            return 0;
        return this._hitCount / total;
    }
    /** Stop cleanup timer and release resources. */
    destroy() {
        if (this.cleanupTimer) {
            clearInterval(this.cleanupTimer);
            this.cleanupTimer = undefined;
        }
        this.store.clear();
        this.currentBytes = 0;
    }
    deleteEntry(key, entry) {
        this.currentBytes -= entry.sizeBytes;
        this.store.delete(key);
    }
    cleanupExpired() {
        const now = Date.now();
        for (const [key, entry] of this.store) {
            if (entry.expiresAt !== undefined && now > entry.expiresAt) {
                this.deleteEntry(key, entry);
            }
        }
    }
}
