// @gao/http/dist/controller.js
/**
 * @gao/http — Controller Route Registry
 *
 * Auto-discovery of routes built from decorated classes.
 * Converts Reflect.metadata into usable router configurations.
 * Supports optional DI Container for constructor injection.
 */
import { CONTROLLER_METADATA_KEY, ROUTES_METADATA_KEY, } from './decorators.js';
import { getValidationSchema, validateData } from './validation.js';
export class ControllerRegistry {
    routes = [];
    container;
    /**
     * Optionally set a DI Container for controller resolution.
     */
    setContainer(container) {
        this.container = container;
    }
    /**
     * Parse a decorated Controller class and extract routes.
     * If a Container is set, uses container.resolveClass() for DI injection.
     * Otherwise falls back to manual instantiation (backward compatible).
     */
    register(ControllerClass) {
        const controllerMeta = Reflect.getMetadata(CONTROLLER_METADATA_KEY, ControllerClass);
        if (!controllerMeta) {
            throw new Error(`Class ${ControllerClass.name} is not decorated with @Controller`);
        }
        const routesMeta = Reflect.getMetadata(ROUTES_METADATA_KEY, ControllerClass) || [];
        // Instantiate via DI Container or manually
        const instance = this.container
            ? this.container.resolveClass(ControllerClass)
            : new ControllerClass();
        for (const route of routesMeta) {
            // Normalize path: handle double slashes
            let fullPath = `${controllerMeta.prefix}${route.path}`.replace(/\/+/g, '/');
            if (fullPath !== '/' && fullPath.endsWith('/')) {
                fullPath = fullPath.slice(0, -1); // remove trailing slash except root
            }
            // Combine controller middlewares with route-specific middlewares
            const combinedMiddlewares = [...controllerMeta.middlewares, ...route.middlewares];
            // Auto-inject validation middleware if @Validate() is present
            const validationSchema = getValidationSchema(instance, route.handlerName);
            if (validationSchema) {
                const validationMw = async (req, _res, next) => {
                    const validated = validateData(validationSchema, req.body);
                    req.setValidated(validated);
                    return next();
                };
                combinedMiddlewares.push(validationMw);
            }
            this.routes.push({
                method: route.method,
                path: fullPath,
                controllerName: ControllerClass.name,
                handlerName: route.handlerName,
                middlewares: combinedMiddlewares,
                // Create a bound handler to ensure 'this' context works inside the controller
                handler: instance[route.handlerName].bind(instance),
            });
        }
    }
    getRoutes() {
        return this.routes;
    }
}
