// @gao/http/dist/health.js
/**
 * @gao/http — Health Check Endpoint
 *
 * Provides a standard `/health` endpoint for Kubernetes readiness/liveness probes.
 */
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
import { Controller, Get } from './decorators.js';
let HealthController = class HealthController {
    // We can inject dependencies here later (like DB connection)
    // For now, let's keep it abstract or allow passing a custom db check function
    dbCheckFn;
    constructor(dbCheckFn) {
        this.dbCheckFn = dbCheckFn;
    }
    async check(_req, res) {
        const result = {
            status: 'up',
            uptime: process.uptime(),
            timestamp: new Date().toISOString(),
            checks: {
                memory: process.memoryUsage(),
            },
        };
        if (this.dbCheckFn) {
            try {
                const dbUp = await this.dbCheckFn();
                result.checks.db = dbUp ? 'up' : 'down';
                if (!dbUp)
                    result.status = 'down';
            }
            catch (e) {
                result.checks.db = 'down';
                result.status = 'down';
            }
        }
        const statusCode = result.status === 'up' ? 200 : 503;
        return res.status(statusCode).json(result);
    }
};
__decorate([
    Get(),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Function, Function]),
    __metadata("design:returntype", Promise)
], HealthController.prototype, "check", null);
HealthController = __decorate([
    Controller('/health'),
    __metadata("design:paramtypes", [Function])
], HealthController);
export { HealthController };
