// @gao/http/dist/session.js
/**
 * @gao/http — Session Management
 *
 * Cookie-based session with pluggable stores.
 */
import { randomUUID } from 'node:crypto';
export class MemorySessionStore {
    store = new Map();
    maxSessions;
    constructor(options) {
        this.maxSessions = options?.maxSessions ?? 10_000;
    }
    async get(id) {
        const entry = this.store.get(id);
        if (!entry)
            return null;
        if (Date.now() > entry.expiresAt) {
            this.store.delete(id);
            return null;
        }
        return { ...entry.data };
    }
    async set(id, data, ttlSeconds) {
        // Evict oldest if at capacity
        if (this.store.size >= this.maxSessions && !this.store.has(id)) {
            const oldest = this.store.keys().next().value;
            if (oldest)
                this.store.delete(oldest);
        }
        this.store.set(id, {
            data: { ...data },
            expiresAt: Date.now() + ttlSeconds * 1000,
        });
    }
    async destroy(id) {
        this.store.delete(id);
    }
    async touch(id, ttlSeconds) {
        const entry = this.store.get(id);
        if (entry) {
            entry.expiresAt = Date.now() + ttlSeconds * 1000;
        }
    }
    /** Get current store size (for debugging) */
    get size() {
        return this.store.size;
    }
}
// ─── Session Instance ──────────────────────────────────────────
export class Session {
    id;
    data = {};
    _isNew = true;
    _isDirty = false;
    constructor(id, initialData) {
        this.id = id;
        if (initialData) {
            this.data = { ...initialData };
            this._isNew = false;
        }
    }
    get(key) {
        return this.data[key];
    }
    set(key, value) {
        this.data[key] = value;
        this._isDirty = true;
    }
    has(key) {
        return key in this.data;
    }
    delete(key) {
        delete this.data[key];
        this._isDirty = true;
    }
    all() {
        return { ...this.data };
    }
    get isNew() {
        return this._isNew;
    }
    get isDirty() {
        return this._isDirty;
    }
}
// ─── Session Manager ───────────────────────────────────────────
export class SessionManager {
    store;
    config;
    constructor(options = {}) {
        this.store = options.store || new MemorySessionStore();
        this.config = {
            cookieName: options.cookieName || 'gao_session',
            ttl: options.ttl || 86400,
            secure: options.secure ?? (process.env.NODE_ENV === 'production'),
            httpOnly: options.httpOnly ?? true,
            sameSite: options.sameSite || 'Lax',
            path: options.path || '/',
            domain: options.domain,
        };
    }
    /**
     * Load or create session from a cookie header string.
     */
    async loadSession(cookieHeader) {
        const sessionId = this.parseCookie(cookieHeader);
        if (sessionId) {
            const data = await this.store.get(sessionId);
            if (data) {
                return new Session(sessionId, data);
            }
        }
        // Create new session
        return new Session(randomUUID());
    }
    /**
     * Save session and return Set-Cookie header value.
     */
    async saveSession(session) {
        if (session.isDirty || session.isNew) {
            await this.store.set(session.id, session.all(), this.config.ttl);
        }
        else {
            await this.store.touch(session.id, this.config.ttl);
        }
        return this.buildSetCookie(session.id);
    }
    /**
     * Destroy a session and return expired Set-Cookie header.
     */
    async destroySession(session) {
        await this.store.destroy(session.id);
        return this.buildSetCookie('', 0); // Expire immediately
    }
    /**
     * Regenerate session (new ID, preserve data).
     */
    async regenerateSession(session) {
        await this.store.destroy(session.id);
        const newSession = new Session(randomUUID());
        for (const [key, value] of Object.entries(session.all())) {
            newSession.set(key, value);
        }
        return newSession;
    }
    // ─── Private Helpers ───────────────────────────────────────
    parseCookie(header) {
        if (!header)
            return null;
        const prefix = `${this.config.cookieName}=`;
        const cookies = header.split(';');
        for (const cookie of cookies) {
            const trimmed = cookie.trim();
            if (trimmed.startsWith(prefix)) {
                return trimmed.slice(prefix.length);
            }
        }
        return null;
    }
    buildSetCookie(value, maxAge) {
        const parts = [`${this.config.cookieName}=${value}`];
        parts.push(`Path=${this.config.path}`);
        parts.push(`Max-Age=${maxAge ?? this.config.ttl}`);
        if (this.config.httpOnly)
            parts.push('HttpOnly');
        if (this.config.secure)
            parts.push('Secure');
        parts.push(`SameSite=${this.config.sameSite}`);
        if (this.config.domain)
            parts.push(`Domain=${this.config.domain}`);
        return parts.join('; ');
    }
}
