// @gao/http/dist/validation.js
/**
 * @gao/http — Request Validation
 *
 * Zod-based request validation with decorator and imperative APIs.
 */
import {} from 'zod';
/**
 * Validation error with field-level messages.
 */
export class HttpValidationError extends Error {
    statusCode = 422;
    errors;
    constructor(zodError) {
        super('Validation failed');
        this.name = 'HttpValidationError';
        this.errors = {};
        for (const issue of zodError.issues) {
            const field = issue.path.join('.');
            if (!this.errors[field]) {
                this.errors[field] = [];
            }
            this.errors[field].push(issue.message);
        }
    }
    toJSON() {
        return {
            error: {
                code: 'VALIDATION_ERROR',
                message: this.message,
                errors: this.errors,
            },
        };
    }
}
/**
 * Validate data against a Zod schema.
 * Returns typed result or throws HttpValidationError.
 */
export function validateData(schema, data) {
    const result = schema.safeParse(data);
    if (!result.success) {
        throw new HttpValidationError(result.error);
    }
    return result.data;
}
/**
 * Metadata key for validation schemas on route handlers.
 */
export const VALIDATION_METADATA_KEY = Symbol('gao:http:validation');
/**
 * Method decorator to validate request body before handler runs.
 *
 * Usage:
 * ```ts
 * @Post('/contacts')
 * @Validate(CreateContactSchema)
 * async create(req: GaoRequest, res: GaoResponse) {
 *   const data = req.validated<CreateContact>();
 * }
 * ```
 */
export function Validate(schema) {
    return (target, propertyKey, _descriptor) => {
        const schemas = Reflect.getMetadata(VALIDATION_METADATA_KEY, target.constructor) || new Map();
        schemas.set(propertyKey, schema);
        Reflect.defineMetadata(VALIDATION_METADATA_KEY, schemas, target.constructor);
    };
}
/**
 * Get the validation schema for a specific handler.
 */
export function getValidationSchema(target, handlerName) {
    const schemas = Reflect.getMetadata(VALIDATION_METADATA_KEY, target.constructor || target);
    return schemas?.get(handlerName);
}
