// @gao/core/dist/cache.js
/**
 * @gao/core — Cache Service
 *
 * In-memory cache with TTL, namespacing, and invalidation.
 * Redis adapter interface for production use.
 * @Cached(ttl) decorator support.
 */
/**
 * In-memory cache adapter (default — no external dependencies).
 */
export class MemoryCacheAdapter {
    store = new Map();
    cleanupInterval;
    constructor(cleanupIntervalMs = 60_000) {
        // Periodic cleanup of expired entries
        this.cleanupInterval = setInterval(() => {
            this.cleanup();
        }, cleanupIntervalMs);
        // Don't prevent Node.js from exiting
        if (this.cleanupInterval.unref) {
            this.cleanupInterval.unref();
        }
    }
    async get(key) {
        const entry = this.store.get(key);
        if (!entry)
            return undefined;
        // Check expiration
        if (entry.expiresAt !== undefined && Date.now() > entry.expiresAt) {
            this.store.delete(key);
            return undefined;
        }
        return entry.value;
    }
    async set(key, value, ttl) {
        this.store.set(key, {
            value,
            expiresAt: ttl ? Date.now() + ttl * 1000 : undefined,
            namespace: this.extractNamespace(key),
        });
    }
    async delete(key) {
        return this.store.delete(key);
    }
    async clear(namespace) {
        if (!namespace) {
            this.store.clear();
            return;
        }
        for (const [key, entry] of this.store) {
            if (entry.namespace === namespace) {
                this.store.delete(key);
            }
        }
    }
    async has(key) {
        const value = await this.get(key);
        return value !== undefined;
    }
    get size() {
        return this.store.size;
    }
    destroy() {
        if (this.cleanupInterval) {
            clearInterval(this.cleanupInterval);
        }
        this.store.clear();
    }
    cleanup() {
        const now = Date.now();
        for (const [key, entry] of this.store) {
            if (entry.expiresAt !== undefined && now > entry.expiresAt) {
                this.store.delete(key);
            }
        }
    }
    extractNamespace(key) {
        const colonIndex = key.indexOf(':');
        return colonIndex > 0 ? key.slice(0, colonIndex) : undefined;
    }
}
/**
 * Cache service with namespace support and decorator.
 */
export class CacheService {
    adapter;
    constructor(adapter = new MemoryCacheAdapter()) {
        this.adapter = adapter;
    }
    async get(key, options) {
        const fullKey = this.buildKey(key, options?.namespace);
        return this.adapter.get(fullKey);
    }
    async set(key, value, options) {
        const fullKey = this.buildKey(key, options?.namespace);
        await this.adapter.set(fullKey, value, options?.ttl);
    }
    /**
     * Get or set: return cached value, or compute and cache it.
     */
    async remember(key, factory, options) {
        const existing = await this.get(key, options);
        if (existing !== undefined)
            return existing;
        const value = await factory();
        await this.set(key, value, options);
        return value;
    }
    async delete(key, options) {
        const fullKey = this.buildKey(key, options?.namespace);
        return this.adapter.delete(fullKey);
    }
    async invalidate(namespace) {
        await this.adapter.clear(namespace);
    }
    async clear() {
        await this.adapter.clear();
    }
    buildKey(key, namespace) {
        return namespace ? `${namespace}:${key}` : key;
    }
}
export function createCacheService(adapter) {
    return new CacheService(adapter);
}
