// @gao/core/dist/errors.js
/**
 * @gao/core — Error System
 *
 * Base error hierarchy with error codes, correlation ID support,
 * and structured error metadata for debugging.
 */
/**
 * Base error class for all GAO Framework errors.
 * Provides error codes, HTTP status mapping, and correlation ID tracking.
 */
export class GaoError extends Error {
    correlationId;
    metadata;
    timestamp;
    constructor(message, options) {
        super(message, { cause: options?.cause });
        this.name = this.constructor.name;
        this.correlationId = options?.correlationId;
        this.metadata = Object.freeze(options?.metadata ?? {});
        this.timestamp = new Date().toISOString();
        // Maintain proper prototype chain
        Object.setPrototypeOf(this, new.target.prototype);
    }
    toJSON() {
        return {
            name: this.name,
            code: this.code,
            message: this.message,
            statusCode: this.statusCode,
            correlationId: this.correlationId,
            metadata: this.metadata,
            timestamp: this.timestamp,
            ...(process.env.NODE_ENV !== 'production' && { stack: this.stack }),
        };
    }
}
// ─── Concrete Error Types ────────────────────────────────────
export class NotFoundError extends GaoError {
    code = 'NOT_FOUND';
    statusCode = 404;
}
export class ValidationError extends GaoError {
    errors;
    code = 'VALIDATION_ERROR';
    statusCode = 422;
    constructor(message, errors, options) {
        super(message, { ...options, metadata: { errors } });
        this.errors = errors;
    }
}
export class UnauthorizedError extends GaoError {
    code = 'UNAUTHORIZED';
    statusCode = 401;
}
export class ForbiddenError extends GaoError {
    code = 'FORBIDDEN';
    statusCode = 403;
}
export class ConflictError extends GaoError {
    code = 'CONFLICT';
    statusCode = 409;
}
export class RateLimitError extends GaoError {
    retryAfter;
    code = 'RATE_LIMIT_EXCEEDED';
    statusCode = 429;
    constructor(message = 'Too many requests', retryAfter = 60, options) {
        super(message, { ...options, metadata: { retryAfter } });
        this.retryAfter = retryAfter;
    }
}
export class InternalError extends GaoError {
    code = 'INTERNAL_ERROR';
    statusCode = 500;
}
export class TimeoutError extends GaoError {
    code = 'TIMEOUT';
    statusCode = 408;
}
/**
 * Type guard: check if an unknown error is a GaoError
 */
export function isGaoError(error) {
    return error instanceof GaoError;
}
