// @gao/core/dist/events.js
/**
 * @gao/core — Event System
 *
 * Typed event emitter for decoupling services:
 * - emit(), on(), once(), off()
 * - Async listeners supported
 * - Type-safe event payloads
 */
export class EventEmitter {
    listeners = new Map();
    /**
     * Subscribe to an event. Returns an unsubscribe handle.
     */
    on(event, listener) {
        if (!this.listeners.has(event)) {
            this.listeners.set(event, new Set());
        }
        const listenersSet = this.listeners.get(event);
        listenersSet?.add(listener);
        return {
            unsubscribe: () => {
                listenersSet?.delete(listener);
            },
        };
    }
    /**
     * Subscribe to an event — fires once then auto-unsubscribes.
     */
    once(event, listener) {
        const wrappedListener = async (payload) => {
            subscription.unsubscribe();
            await listener(payload);
        };
        const subscription = this.on(event, wrappedListener);
        return subscription;
    }
    /**
     * Remove a specific listener from an event.
     */
    off(event, listener) {
        this.listeners.get(event)?.delete(listener);
    }
    /**
     * Emit an event to all listeners. Async listeners are awaited.
     */
    async emit(event, payload) {
        const listenersSet = this.listeners.get(event);
        if (!listenersSet || listenersSet.size === 0)
            return;
        const promises = [];
        for (const listener of listenersSet) {
            const result = listener(payload);
            if (result instanceof Promise) {
                promises.push(result);
            }
        }
        if (promises.length > 0) {
            await Promise.all(promises);
        }
    }
    /**
     * Remove all listeners for a specific event, or all events.
     */
    removeAll(event) {
        if (event) {
            this.listeners.delete(event);
        }
        else {
            this.listeners.clear();
        }
    }
    /**
     * Get listener count for a specific event.
     */
    listenerCount(event) {
        return this.listeners.get(event)?.size ?? 0;
    }
    /**
     * Get all registered event names.
     */
    eventNames() {
        return [...this.listeners.keys()];
    }
}
export function createEventEmitter() {
    return new EventEmitter();
}
