// @gao/core/dist/plugin-marketplace.js
/**
 * @gao/core — Plugin Marketplace
 *
 * Orchestrates plugin lifecycle for trusted and sandboxed plugins:
 * - Install: load trusted plugins in-process with ScopedContainer, worker plugins in WorkerSandbox.
 * - Invoke: broadcast hooks to all installed plugins.
 * - Uninstall: cleanup and terminate workers.
 */
import { PermissionSet } from './permission.js';
import { ScopedContainer } from './scoped-container.js';
import { WorkerSandbox } from './worker-sandbox.js';
import { validateManifest } from './plugin-manifest.js';
export class PluginMarketplace {
    container;
    workerHostPath;
    plugins = new Map();
    constructor(container, workerHostPath) {
        this.container = container;
        this.workerHostPath = workerHostPath;
    }
    /**
     * Install a plugin from its manifest.
     * - "trusted" sandbox → ScopedContainer + direct import
     * - "worker" sandbox → WorkerSandbox with resource limits
     * - "wasm" sandbox → Placeholder for future WASM support
     */
    async install(manifest) {
        const [valid, errorOrManifest] = validateManifest(manifest);
        if (!valid) {
            throw new Error(`Invalid plugin manifest: ${errorOrManifest}`);
        }
        if (this.plugins.has(manifest.name)) {
            throw new Error(`Plugin "${manifest.name}" is already installed`);
        }
        if (manifest.sandbox === 'trusted') {
            const permissions = new PermissionSet(manifest.permissions);
            const scopedContainer = new ScopedContainer(this.container, permissions, manifest.name);
            this.plugins.set(manifest.name, {
                manifest,
                sandbox: scopedContainer,
                type: 'trusted',
            });
        }
        else if (manifest.sandbox === 'worker') {
            const sandbox = new WorkerSandbox({
                maxHeapMB: manifest.budget?.maxHeapMB ?? 128,
            });
            const hostPath = this.workerHostPath ?? new URL('./plugin-worker-host.js', import.meta.url).pathname;
            await sandbox.load(manifest, hostPath);
            this.plugins.set(manifest.name, {
                manifest,
                sandbox,
                type: 'worker',
            });
        }
        else if (manifest.sandbox === 'wasm') {
            throw new Error('WASM sandbox is not yet supported. Use "trusted" or "worker".');
        }
    }
    /**
     * Invoke a hook on all installed plugins that declared it.
     * Returns a map of plugin name → result.
     */
    async invokeHook(hook, data) {
        const results = new Map();
        for (const [name, plugin] of this.plugins) {
            if (plugin.manifest.hooks && !plugin.manifest.hooks.includes(hook)) {
                continue; // Plugin didn't declare this hook
            }
            try {
                if (plugin.type === 'worker' && plugin.sandbox instanceof WorkerSandbox) {
                    const result = await plugin.sandbox.invoke(hook, data);
                    results.set(name, result);
                }
                else {
                    // For trusted plugins, hook invocation is handled by the plugin itself
                    results.set(name, undefined);
                }
            }
            catch (err) {
                results.set(name, { error: err.message });
            }
        }
        return results;
    }
    /** Uninstall a plugin by name, terminating its sandbox if necessary. */
    async uninstall(name) {
        const plugin = this.plugins.get(name);
        if (!plugin)
            return false;
        if (plugin.type === 'worker' && plugin.sandbox instanceof WorkerSandbox) {
            await plugin.sandbox.terminate();
        }
        this.plugins.delete(name);
        return true;
    }
    /** List all installed plugins. */
    list() {
        return [...this.plugins.values()];
    }
    /** Get an installed plugin by name. */
    get(name) {
        return this.plugins.get(name);
    }
    /** Number of installed plugins. */
    get size() {
        return this.plugins.size;
    }
    /** Uninstall all plugins and terminate all workers. */
    async shutdown() {
        for (const [name] of this.plugins) {
            await this.uninstall(name);
        }
    }
}
