// @gao/core/dist/worker-sandbox.js
/**
 * @gao/core — Worker Sandbox
 *
 * Runs plugin code in an isolated Node.js Worker Thread with:
 * - Configurable V8 resource limits (maxOldGenerationSizeMb)
 * - Message-passing based RPC (invoke hook → response/error)
 * - Automatic timeout enforcement per invocation
 * - Graceful termination and crash recovery
 */
import { Worker } from 'node:worker_threads';
export class WorkerSandbox {
    options;
    worker;
    pendingCalls = new Map();
    callId = 0;
    _isRunning = false;
    _manifest;
    constructor(options = {}) {
        this.options = options;
    }
    /** Load a plugin in a Worker Thread with resource limits. */
    async load(manifest, hostPath) {
        this._manifest = manifest;
        const maxHeapMB = manifest.budget?.maxHeapMB ?? this.options.maxHeapMB ?? 128;
        this.worker = new Worker(hostPath, {
            workerData: {
                manifest,
                pluginPath: manifest.main,
            },
            resourceLimits: {
                maxOldGenerationSizeMb: maxHeapMB,
                maxYoungGenerationSizeMb: Math.max(8, Math.floor(maxHeapMB / 4)),
            },
        });
        this._isRunning = true;
        this.worker.on('message', (msg) => {
            const pending = this.pendingCalls.get(msg.callId);
            if (!pending)
                return;
            clearTimeout(pending.timer);
            this.pendingCalls.delete(msg.callId);
            if (msg.error) {
                pending.reject(new Error(`Plugin error: ${msg.error}`));
            }
            else {
                pending.resolve(msg.result);
            }
        });
        this.worker.on('error', (err) => {
            this.rejectAllPending(err);
        });
        this.worker.on('exit', (code) => {
            this._isRunning = false;
            if (code !== 0) {
                this.rejectAllPending(new Error(`Worker exited with code ${code}`));
            }
        });
        // Wait for worker to report ready
        await new Promise((resolve, reject) => {
            const readyTimeout = setTimeout(() => {
                reject(new Error('Worker did not become ready within timeout'));
            }, 10000);
            const onMessage = (msg) => {
                if (msg.type === 'ready') {
                    clearTimeout(readyTimeout);
                    this.worker?.off('message', onMessage);
                    resolve();
                }
            };
            this.worker.on('message', onMessage);
        });
    }
    /**
     * Invoke a hook on the sandboxed plugin.
     * Returns the result or throws on error/timeout.
     */
    async invoke(hook, data, timeoutMs) {
        if (!this.worker || !this._isRunning) {
            throw new Error('Worker is not running');
        }
        const id = String(++this.callId);
        const timeout = timeoutMs ?? this.options.defaultTimeoutMs ?? 5000;
        return new Promise((resolve, reject) => {
            const timer = setTimeout(() => {
                this.pendingCalls.delete(id);
                reject(new Error(`Plugin hook "${hook}" timed out after ${timeout}ms`));
            }, timeout);
            this.pendingCalls.set(id, { resolve, reject, timer });
            this.worker.postMessage({ callId: id, hook, data });
        });
    }
    /** Gracefully terminate the worker. */
    async terminate() {
        if (!this.worker)
            return;
        this.rejectAllPending(new Error('Worker terminated'));
        await this.worker.terminate();
        this.worker = undefined;
        this._isRunning = false;
    }
    /** Whether the worker is currently running. */
    get isRunning() {
        return this._isRunning;
    }
    /** The loaded plugin manifest. */
    get manifest() {
        return this._manifest;
    }
    rejectAllPending(error) {
        for (const [id, pending] of this.pendingCalls) {
            clearTimeout(pending.timer);
            pending.reject(error);
            this.pendingCalls.delete(id);
        }
    }
}
