// @gao/security/dist/index.js
/**
 * @gao/security — Barrel Export
 *
 * Public API surface for @gao/security.
 */
// ─── Middleware ───────────────────────────────────────────────
export { helmet } from './middleware/helmet.js';
export { cors } from './middleware/cors.js';
export { csrf, generateCsrfSecret, createCsrfToken, verifyCsrfToken } from './middleware/csrf.js';
export { rateLimiter, MemoryRateLimiterStore } from './middleware/rate-limiter.js';
export { validate } from './middleware/validator.js';
export { xssGuard, htmlEncode, sanitizeHtml, generateCspNonce } from './middleware/xss-guard.js';
export { ddosShield } from './middleware/ddos-shield.js';
export { secureUpload, sanitizeFilename } from './middleware/upload.js';
export { captchaShield, TurnstileProvider, RecaptchaV3Provider } from './middleware/captcha.js';
// ─── Crypto ───────────────────────────────────────────────────
export { hashPassword, verifyPassword, needsRehash } from './crypto/hashing.js';
export { encrypt, decrypt, deriveKey } from './crypto/aes256.js';
export { Encrypted, encryptFields, decryptFields, ENCRYPTED_FIELDS_META_KEY, } from './crypto/field-encrypt.js';
// ─── Auth ─────────────────────────────────────────────────────
export { JwtService, MemoryRevocationAdapter } from './auth/jwt.js';
export { RbacEngine } from './auth/rbac.js';
export { guard } from './auth/guard.js';
