// @gao/core/dist/redis.js
/**
 * @gao/core — Redis Cache Adapter & Client Factory
 *
 * Implements CacheAdapter using ioredis (optional peer dependency).
 * All keys are prefixed with a configurable namespace to avoid collisions.
 */
/**
 * Redis-backed cache adapter implementing CacheAdapter.
 */
export class RedisCacheAdapter {
    client;
    prefix;
    constructor(client, options) {
        this.client = client;
        this.prefix = options?.prefix ?? 'gao:cache:';
    }
    prefixKey(key) {
        return `${this.prefix}${key}`;
    }
    async get(key) {
        const data = await this.client.get(this.prefixKey(key));
        if (data === null)
            return undefined;
        try {
            return JSON.parse(data);
        }
        catch {
            return data;
        }
    }
    async set(key, value, ttl) {
        const serialized = JSON.stringify(value);
        const prefixed = this.prefixKey(key);
        if (ttl && ttl > 0) {
            await this.client.set(prefixed, serialized, 'EX', ttl);
        }
        else {
            await this.client.set(prefixed, serialized);
        }
    }
    async delete(key) {
        const result = await this.client.del(this.prefixKey(key));
        return result > 0;
    }
    async has(key) {
        const result = await this.client.exists(this.prefixKey(key));
        return result > 0;
    }
    async clear(namespace) {
        const pattern = namespace
            ? `${this.prefix}${namespace}:*`
            : `${this.prefix}*`;
        const keys = await this.client.keys(pattern);
        if (keys.length > 0) {
            await this.client.del(...keys);
        }
    }
    /**
     * Disconnect the Redis client.
     */
    async destroy() {
        await this.client.quit();
    }
}
/**
 * Create a Redis client using ioredis.
 * Requires `ioredis` to be installed as a peer dependency.
 *
 * @throws Error if ioredis is not installed
 */
export async function createRedisClient(config = {}) {
    let Redis;
    try {
        // @ts-ignore — ioredis is an optional peer dependency
        const mod = await import('ioredis');
        Redis = mod.default || mod;
    }
    catch {
        throw new Error('ioredis is required for Redis support. Install it: pnpm add ioredis');
    }
    if (config.url) {
        return new Redis(config.url);
    }
    return new Redis({
        host: config.host ?? '127.0.0.1',
        port: config.port ?? 6379,
        password: config.password,
        db: config.db ?? 0,
        keyPrefix: config.keyPrefix,
    });
}
