// @gao/http/dist/request.js
/**
 * @gao/http — GaoRequest
 *
 * A framework-specific wrapper around the standard Web API Request object.
 * Provides easy access to query parameters, routing params, parsed body,
 * custom headers like Correlation-ID, IP address, and user context.
 */
import { randomUUID } from 'node:crypto';
export class GaoRequest {
    native;
    method;
    url;
    ip;
    correlationId;
    // Parsed data
    body = {};
    query = {};
    params = {};
    // User Context (to be populated by auth middlewares)
    user;
    // Session (populated by sessionMiddleware)
    session;
    // Flash messages (populated by flashMiddleware)
    flash;
    // Validated request body (populated by validation middleware)
    _validated;
    constructor(req, clientIp = '127.0.0.1') {
        this.native = req;
        this.method = req.method;
        this.url = new URL(req.url);
        // Extract IP (using provided, checking headers fallback if needed, like X-Forwarded-For)
        const forwarded = req.headers.get('x-forwarded-for');
        this.ip = forwarded ? forwarded.split(',')[0].trim() : clientIp;
        // Correlation ID
        this.correlationId = req.headers.get('x-correlation-id') || randomUUID();
        // Parse query params into a simple object
        const q = {};
        for (const [key, value] of this.url.searchParams.entries()) {
            q[key] = value;
        }
        this.query = q;
    }
    /**
     * Initializes the request body parsing.
     * Call this before accessing `req.body` if parsing is required.
     */
    async parseBody() {
        if (this.method === 'GET' || this.method === 'HEAD') {
            return; // No body
        }
        const contentType = this.native.headers.get('content-type') || '';
        try {
            if (contentType.includes('application/json')) {
                this.body = await this.native.clone().json();
            }
            else if (contentType.includes('application/x-www-form-urlencoded')) {
                const text = await this.native.clone().text();
                const searchParams = new URLSearchParams(text);
                const formData = {};
                for (const [key, value] of searchParams.entries()) {
                    formData[key] = value;
                }
                this.body = formData;
            }
            else if (contentType.includes('multipart/form-data')) {
                // Return Native FormData object as the body, user can iterate easily
                this.body = (await this.native.clone().formData());
            }
            else {
                // Fallback to text
                this.body = (await this.native.clone().text());
            }
        }
        catch (e) {
            // Unparseable body
            this.body = {};
        }
    }
    /**
     * Helper to get a specific header.
     */
    header(name) {
        return this.native.headers.get(name);
    }
    /**
     * Get the validated request body (set by @Validate decorator middleware).
     */
    validated() {
        if (this._validated === undefined) {
            throw new Error('No validated data available. Did you apply @Validate() decorator?');
        }
        return this._validated;
    }
    /**
     * Set validated data (called by validation middleware — not for direct use).
     */
    setValidated(data) {
        this._validated = data;
    }
}
