// @gao/http/dist/response.js
/**
 * @gao/http — GaoResponse
 *
 * Builder pattern for standard Web API Response objects.
 * Allows fluent setting of headers, status, and payload.
 * Also enforces the GAO generic success/error envelope shape.
 */
export class GaoResponse {
    _status = 200;
    _headers = new Headers();
    _body = null;
    /**
     * Set the HTTP status code
     */
    status(code) {
        this._status = code;
        return this;
    }
    /**
     * Add or set an HTTP header
     */
    header(name, value) {
        this._headers.set(name, value);
        return this;
    }
    /**
     * Append to an HTTP header
     */
    appendHeader(name, value) {
        this._headers.append(name, value);
        return this;
    }
    /**
     * Sets the response body to JSON, wrapped in the standard envelope.
     */
    json(data, meta) {
        this.header('Content-Type', 'application/json; charset=utf-8');
        const envelope = { data };
        if (meta) {
            envelope.meta = meta;
        }
        this._body = JSON.stringify(envelope);
        return this.build();
    }
    /**
     * Send an error response following the envelope scheme.
     */
    error(status, code, message, details) {
        this.status(status);
        this.header('Content-Type', 'application/json; charset=utf-8');
        const envelope = {
            error: { code, message, details },
        };
        this._body = JSON.stringify(envelope);
        return this.build();
    }
    /**
     * Sends plain text response
     */
    text(text) {
        this.header('Content-Type', 'text/plain; charset=utf-8');
        this._body = text;
        return this.build();
    }
    /**
     * Sends a raw HTML response
     */
    html(html) {
        this.header('Content-Type', 'text/html; charset=utf-8');
        this._body = html;
        return this.build();
    }
    /**
     * Returns a redirect response
     */
    redirect(url, status = 302) {
        this.status(status);
        this.header('Location', url);
        return this.build();
    }
    /**
     * Sends a raw stream response
     */
    stream(stream, contentType = 'application/octet-stream') {
        this.header('Content-Type', contentType);
        this._body = stream;
        return this.build();
    }
    /**
     * Return empty response (204 No Content)
     */
    empty() {
        this.status(204);
        this._body = null;
        return this.build();
    }
    /**
     * Reset the response to its initial state for object pool reuse.
     * Clears status, headers, and body.
     */
    reset() {
        this._status = 200;
        this._headers = new Headers();
        this._body = null;
    }
    /**
     * Constructs the final native Web Request Response object.
     * Normally called internally by terminators like .json(), .html() etc.
     */
    build() {
        return new Response(this._body, {
            status: this._status,
            headers: this._headers,
        });
    }
}
