// @gao/http/dist/session-redis.js
/**
 * @gao/http — Redis Session Store
 *
 * Redis-backed session store implementing SessionStore interface.
 * Uses ioredis (optional peer dependency) for persistence.
 */
/**
 * Redis-backed session store for production use.
 *
 * Usage:
 * ```ts
 * import { createRedisClient } from '@gao/core';
 * import { RedisSessionStore, sessionMiddleware } from '@gao/http';
 *
 * const redis = await createRedisClient({ host: '127.0.0.1' });
 * const store = new RedisSessionStore(redis);
 *
 * app.use(sessionMiddleware({ store }));
 * ```
 */
export class RedisSessionStore {
    client;
    prefix;
    constructor(client, options) {
        this.client = client;
        this.prefix = options?.prefix ?? 'gao:session:';
    }
    prefixKey(id) {
        return `${this.prefix}${id}`;
    }
    async get(id) {
        const data = await this.client.get(this.prefixKey(id));
        if (data === null)
            return null;
        try {
            return JSON.parse(data);
        }
        catch {
            return null;
        }
    }
    async set(id, data, ttlSeconds) {
        const serialized = JSON.stringify(data);
        await this.client.set(this.prefixKey(id), serialized, 'EX', ttlSeconds);
    }
    async destroy(id) {
        await this.client.del(this.prefixKey(id));
    }
    async touch(id, ttlSeconds) {
        await this.client.expire(this.prefixKey(id), ttlSeconds);
    }
}
